// @tambo-ai/typescript-sdk/src/resources/threads/messages.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import { APIResource } from '../../core/resource';
import * as ThreadsAPI from './threads';
import { APIPromise } from '../../core/api-promise';
import { RequestOptions } from '../../internal/request-options';
import { path } from '../../internal/utils/path';

export class Messages extends APIResource {
  /**
   * List messages in a thread. Supports cursor-based pagination and ordering.
   *
   * @example
   * ```ts
   * const messages = await client.threads.messages.list(
   *   'thr_abc123xyz',
   * );
   * ```
   */
  list(
    threadID: string,
    query: MessageListParams | null | undefined = {},
    options?: RequestOptions,
  ): APIPromise<MessageListResponse> {
    return this._client.get(path`/v1/threads/${threadID}/messages`, { query, ...options });
  }

  /**
   * Get a specific message by ID from a thread.
   *
   * @example
   * ```ts
   * const message = await client.threads.messages.get(
   *   'msg_xyz789abc',
   *   { threadId: 'thr_abc123xyz' },
   * );
   * ```
   */
  get(messageID: string, params: MessageGetParams, options?: RequestOptions): APIPromise<MessageGetResponse> {
    const { threadId } = params;
    return this._client.get(path`/v1/threads/${threadId}/messages/${messageID}`, options);
  }
}

export interface MessageListResponse {
  /**
   * Whether there are more results
   */
  hasMore: boolean;

  /**
   * List of messages in the thread
   */
  messages: Array<MessageListResponse.Message>;

  /**
   * Cursor for the next page of results
   */
  nextCursor?: string;
}

export namespace MessageListResponse {
  export interface Message {
    /**
     * Unique identifier for this message
     */
    id: string;

    /**
     * Content blocks in this message
     */
    content: Array<
      | ThreadsAPI.TextContent
      | ThreadsAPI.ResourceContent
      | ThreadsAPI.ToolUseContent
      | ThreadsAPI.ToolResultContent
      | ThreadsAPI.ComponentContent
    >;

    /**
     * Message role
     */
    role: 'user' | 'assistant' | 'system';

    /**
     * When the message was created (ISO 8601)
     */
    createdAt?: string;

    /**
     * Additional metadata
     */
    metadata?: unknown;
  }
}

export interface MessageGetResponse {
  /**
   * Unique identifier for this message
   */
  id: string;

  /**
   * Content blocks in this message
   */
  content: Array<
    | ThreadsAPI.TextContent
    | ThreadsAPI.ResourceContent
    | ThreadsAPI.ToolUseContent
    | ThreadsAPI.ToolResultContent
    | ThreadsAPI.ComponentContent
  >;

  /**
   * Message role
   */
  role: 'user' | 'assistant' | 'system';

  /**
   * When the message was created (ISO 8601)
   */
  createdAt?: string;

  /**
   * Additional metadata
   */
  metadata?: unknown;
}

export interface MessageListParams {
  /**
   * Cursor for pagination
   */
  cursor?: string;

  /**
   * Maximum number of messages to return
   */
  limit?: string;

  /**
   * Sort order: 'asc' for oldest first, 'desc' for newest first
   */
  order?: 'asc' | 'desc';
}

export interface MessageGetParams {
  /**
   * Thread ID
   */
  threadId: string;
}

export declare namespace Messages {
  export {
    type MessageListResponse as MessageListResponse,
    type MessageGetResponse as MessageGetResponse,
    type MessageListParams as MessageListParams,
    type MessageGetParams as MessageGetParams,
  };
}
