// @tambo-ai/typescript-sdk/src/resources/threads/state.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import { APIResource } from '../../core/resource';
import { APIPromise } from '../../core/api-promise';
import { RequestOptions } from '../../internal/request-options';
import { path } from '../../internal/utils/path';

export class State extends APIResource {
  /**
   * Update the state of a component in a thread. Supports both full replacement and
   * JSON Patch operations. Thread must not have an active run.
   *
   * @example
   * ```ts
   * const response = await client.threads.state.updateState(
   *   'comp_xyz789abc',
   *   { threadId: 'thr_abc123xyz' },
   * );
   * ```
   */
  updateState(
    componentID: string,
    params: StateUpdateStateParams,
    options?: RequestOptions,
  ): APIPromise<StateUpdateStateResponse> {
    const { threadId, ...body } = params;
    return this._client.post(path`/v1/threads/${threadId}/components/${componentID}/state`, {
      body,
      ...options,
    });
  }
}

export interface StateUpdateStateResponse {
  /**
   * New complete state of the component
   */
  state: unknown;
}

export interface StateUpdateStateParams {
  /**
   * Path param: Thread ID
   */
  threadId: string;

  /**
   * Body param: JSON Patch operations to apply to current state
   */
  patch?: Array<StateUpdateStateParams.Patch>;

  /**
   * Body param: Full replacement state object
   */
  state?: unknown;
}

export namespace StateUpdateStateParams {
  export interface Patch {
    /**
     * Operation type
     */
    op: 'add' | 'remove' | 'replace' | 'move' | 'copy' | 'test';

    /**
     * JSON Pointer path to the target location
     */
    path: string;

    /**
     * Source path for move and copy operations
     */
    from?: string;

    /**
     * Value to add, replace, or test (required for add, replace, test)
     */
    value?: unknown;
  }
}

export declare namespace State {
  export {
    type StateUpdateStateResponse as StateUpdateStateResponse,
    type StateUpdateStateParams as StateUpdateStateParams,
  };
}
