// @tambo-ai/react/dist/setupTests.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("@testing-library/jest-dom");
const util_1 = require("util");
const web_1 = require("stream/web");
// Mock Date.now() to return a fixed timestamp for consistent testing
const mockDate = new Date(2025, 0, 5, 12, 32, 58, 936);
global.Date.now = jest.fn(() => mockDate.getTime());
// Add TextEncoder/TextDecoder polyfills for Node.js test environment
Object.assign(global, { TextEncoder: util_1.TextEncoder, TextDecoder: util_1.TextDecoder });
// Add Web Streams API polyfills for jsdom environment (used by @modelcontextprotocol/sdk)
Object.assign(global, { TransformStream: web_1.TransformStream, ReadableStream: web_1.ReadableStream, WritableStream: web_1.WritableStream });
// Mock react-media-recorder for tests (browser APIs not available in Node.js)
jest.mock("react-media-recorder", () => ({
    useReactMediaRecorder: () => ({
        status: "idle",
        startRecording: jest.fn(),
        stopRecording: jest.fn(),
        mediaBlobUrl: null,
        error: null,
    }),
}));
