// @tambo-ai/typescript-sdk/resources/threads/runs.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Runs = void 0;
const resource_1 = require("../../core/resource.js");
const headers_1 = require("../../internal/headers.js");
const path_1 = require("../../internal/utils/path.js");
class Runs extends resource_1.APIResource {
    /**
     * Creates a new thread and immediately starts a streaming run. Returns an SSE
     * stream of AG-UI events.
     *
     * @example
     * ```ts
     * const run = await client.threads.runs.create({
     *   message: {
     *     content: [{ text: 'Hello, world!', type: 'text' }],
     *     role: 'user',
     *   },
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/v1/threads/runs', {
            body,
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: 'text/event-stream' }, options?.headers]),
            stream: true,
        });
    }
    /**
     * Explicitly cancel a running run. Note: closing the SSE connection also cancels
     * the run.
     *
     * @example
     * ```ts
     * const run = await client.threads.runs.delete(
     *   'run_xyz789abc',
     *   { threadId: 'thr_abc123xyz' },
     * );
     * ```
     */
    delete(runID, params, options) {
        const { threadId } = params;
        return this._client.delete((0, path_1.path) `/v1/threads/${threadId}/runs/${runID}`, options);
    }
    /**
     * Starts a streaming run on an existing thread. Returns an SSE stream of AG-UI
     * events.
     *
     * @example
     * ```ts
     * const response = await client.threads.runs.run(
     *   'thr_abc123xyz',
     *   {
     *     message: {
     *       content: [{ text: 'Hello, world!', type: 'text' }],
     *       role: 'user',
     *     },
     *   },
     * );
     * ```
     */
    run(threadID, body, options) {
        return this._client.post((0, path_1.path) `/v1/threads/${threadID}/runs`, {
            body,
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: 'text/event-stream' }, options?.headers]),
            stream: true,
        });
    }
}
exports.Runs = Runs;
