// @tambo-ai/typescript-sdk/resources/threads/threads.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../core/resource.mjs";
import * as MessagesAPI from "./messages.mjs";
import { Messages, } from "./messages.mjs";
import * as RunsAPI from "./runs.mjs";
import { Runs, } from "./runs.mjs";
import * as StateAPI from "./state.mjs";
import { State } from "./state.mjs";
import { buildHeaders } from "../../internal/headers.mjs";
import { path } from "../../internal/utils/path.mjs";
export class Threads extends APIResource {
    constructor() {
        super(...arguments);
        this.state = new StateAPI.State(this._client);
        this.messages = new MessagesAPI.Messages(this._client);
        this.runs = new RunsAPI.Runs(this._client);
    }
    /**
     * Create a new empty thread. Note: initialMessages is not supported yet; create
     * the thread first, then add messages via runs/message endpoints.
     *
     * @example
     * ```ts
     * const thread = await client.threads.create();
     * ```
     */
    create(body, options) {
        return this._client.post('/v1/threads', { body, ...options });
    }
    /**
     * Get a thread by ID with all its messages. The thread must belong to the
     * authenticated project.
     *
     * @example
     * ```ts
     * const thread = await client.threads.retrieve(
     *   'thr_abc123xyz',
     * );
     * ```
     */
    retrieve(threadID, query = {}, options) {
        return this._client.get(path `/v1/threads/${threadID}`, { query, ...options });
    }
    /**
     * List all threads for the authenticated project. Supports cursor-based pagination
     * and filtering by user key.
     *
     * @example
     * ```ts
     * const threads = await client.threads.list();
     * ```
     */
    list(query = {}, options) {
        return this._client.get('/v1/threads', { query, ...options });
    }
    /**
     * Delete a thread and all its messages. This action cannot be undone.
     *
     * @example
     * ```ts
     * await client.threads.delete('thr_abc123xyz');
     * ```
     */
    delete(threadID, options) {
        return this._client.delete(path `/v1/threads/${threadID}`, {
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
}
Threads.State = State;
Threads.Messages = Messages;
Threads.Runs = Runs;
