// @tambo-ai/typescript-sdk/resources/threads/threads.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Threads = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const MessagesAPI = tslib_1.__importStar(require("./messages.js"));
const messages_1 = require("./messages.js");
const RunsAPI = tslib_1.__importStar(require("./runs.js"));
const runs_1 = require("./runs.js");
const StateAPI = tslib_1.__importStar(require("./state.js"));
const state_1 = require("./state.js");
const headers_1 = require("../../internal/headers.js");
const path_1 = require("../../internal/utils/path.js");
class Threads extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.state = new StateAPI.State(this._client);
        this.messages = new MessagesAPI.Messages(this._client);
        this.runs = new RunsAPI.Runs(this._client);
    }
    /**
     * Create a new empty thread. Note: initialMessages is not supported yet; create
     * the thread first, then add messages via runs/message endpoints.
     *
     * @example
     * ```ts
     * const thread = await client.threads.create();
     * ```
     */
    create(body, options) {
        return this._client.post('/v1/threads', { body, ...options });
    }
    /**
     * Get a thread by ID with all its messages. The thread must belong to the
     * authenticated project.
     *
     * @example
     * ```ts
     * const thread = await client.threads.retrieve(
     *   'thr_abc123xyz',
     * );
     * ```
     */
    retrieve(threadID, query = {}, options) {
        return this._client.get((0, path_1.path) `/v1/threads/${threadID}`, { query, ...options });
    }
    /**
     * List all threads for the authenticated project. Supports cursor-based pagination
     * and filtering by user key.
     *
     * @example
     * ```ts
     * const threads = await client.threads.list();
     * ```
     */
    list(query = {}, options) {
        return this._client.get('/v1/threads', { query, ...options });
    }
    /**
     * Delete a thread and all its messages. This action cannot be undone.
     *
     * @example
     * ```ts
     * await client.threads.delete('thr_abc123xyz');
     * ```
     */
    delete(threadID, options) {
        return this._client.delete((0, path_1.path) `/v1/threads/${threadID}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
}
exports.Threads = Threads;
Threads.State = state_1.State;
Threads.Messages = messages_1.Messages;
Threads.Runs = runs_1.Runs;
