// @tambo-ai/react/dist/mcp/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTamboElicitationContext = exports.useTamboMcpServers = exports.useTamboMcpElicitation = exports.TamboMcpProvider = exports.isMcpResourceEntry = exports.useTamboMcpResourceList = exports.useTamboMcpResource = exports.useTamboMcpPromptList = exports.useTamboMcpPrompt = exports.MCPTransport = void 0;
/**
 * Entry point for MCP (Model Context Protocol) support in the React SDK.
 *
 * Note: The `@modelcontextprotocol/sdk` is included automatically with `@tambo-ai/react`.
 * If you use features that require schema validation (like component props schemas),
 * you'll need to install `zod` and `zod-to-json-schema` as optional peer dependencies.
 * See the React SDK README for the recommended version ranges.
 */
var mcp_client_1 = require("./mcp-client");
Object.defineProperty(exports, "MCPTransport", { enumerable: true, get: function () { return mcp_client_1.MCPTransport; } });
var mcp_hooks_1 = require("./mcp-hooks");
Object.defineProperty(exports, "useTamboMcpPrompt", { enumerable: true, get: function () { return mcp_hooks_1.useTamboMcpPrompt; } });
Object.defineProperty(exports, "useTamboMcpPromptList", { enumerable: true, get: function () { return mcp_hooks_1.useTamboMcpPromptList; } });
Object.defineProperty(exports, "useTamboMcpResource", { enumerable: true, get: function () { return mcp_hooks_1.useTamboMcpResource; } });
Object.defineProperty(exports, "useTamboMcpResourceList", { enumerable: true, get: function () { return mcp_hooks_1.useTamboMcpResourceList; } });
Object.defineProperty(exports, "isMcpResourceEntry", { enumerable: true, get: function () { return mcp_hooks_1.isMcpResourceEntry; } });
var tambo_mcp_provider_1 = require("./tambo-mcp-provider");
Object.defineProperty(exports, "TamboMcpProvider", { enumerable: true, get: function () { return tambo_mcp_provider_1.TamboMcpProvider; } });
Object.defineProperty(exports, "useTamboMcpElicitation", { enumerable: true, get: function () { return tambo_mcp_provider_1.useTamboMcpElicitation; } });
Object.defineProperty(exports, "useTamboMcpServers", { enumerable: true, get: function () { return tambo_mcp_provider_1.useTamboMcpServers; } });
Object.defineProperty(exports, "useTamboElicitationContext", { enumerable: true, get: function () { return tambo_mcp_provider_1.useTamboElicitationContext; } });
