// @tambo-ai/react/esm/hooks/use-current-message.js
"use client";
import React, { createContext, useContext } from "react";
export const TamboMessageContext = createContext(null);
/**
 * Wraps all components, so that they can find what message they are in.
 * Also supports optional interactable metadata for components wrapped with withInteractable.
 * @param props - props for the TamboMessageProvider
 * @param props.children - The children to wrap
 * @param props.message - The message object
 * @param props.interactableMetadata - Optional interactable component metadata
 * @returns The wrapped component
 */
export const TamboMessageProvider = ({ children, message, interactableMetadata, }) => {
    // Merge interactable metadata into message if provided
    const enhancedMessage = interactableMetadata
        ? { ...message, interactableMetadata }
        : message;
    // Use a unique key={...} to force a re-render when the message changes - this
    // make sure that if the rendered component is swapped into a tree (like if
    // you always show the last rendered component) then the state/etc is correct
    return (React.createElement(TamboMessageContext.Provider, { value: enhancedMessage, key: message.id }, children));
};
/**
 * Wraps a component with a TamboMessageProvider - this allows the provider
 * to be used outside of a TSX file
 * @param children - The children to wrap
 * @param message - The message object
 * @param interactableMetadata - Optional interactable metadata
 * @returns The wrapped component
 */
export function wrapWithTamboMessageProvider(children, message, interactableMetadata) {
    return (React.createElement(TamboMessageProvider, { message: message, interactableMetadata: interactableMetadata }, children));
}
/**
 * Hook used inside a component wrapped with TamboMessageProvider, to get
 * the current message.
 * @returns The current message that is used to render the component
 */
export const useTamboCurrentMessage = () => {
    const message = useContext(TamboMessageContext);
    if (!message) {
        throw new Error("useTamboCurrentMessage must be used within a TamboMessageProvider");
    }
    return message;
};
/**
 * Hook to access the current component information from the message context.
 * Provides a unified interface for both AI-generated and interactable components.
 *
 * **Use this hook when you need component metadata regardless of the context.**
 * @returns Component info including name, props, and interactable metadata if available, or null if used outside TamboMessageProvider
 * @example
 * ```tsx
 * function MyInlineEditor() {
 *   const component = useTamboCurrentComponent();
 *
 *   if (!component) return null; // Not inside a component
 *
 *   return (
 *     <div>
 *       Editing: {component.componentName}
 *       {component.interactableId && <span>ID: {component.interactableId}</span>}
 *     </div>
 *   );
 * }
 * ```
 */
export const useTamboCurrentComponent = () => {
    const message = useContext(TamboMessageContext);
    if (!message) {
        return null;
    }
    return {
        componentName: message.interactableMetadata?.componentName ??
            message.component?.componentName ??
            undefined,
        props: message.component?.props,
        interactableId: message.interactableMetadata?.id ?? undefined,
        description: message.interactableMetadata?.description ?? undefined,
        threadId: message.threadId,
    };
};
