// @tambo-ai/react/esm/v1/providers/tambo-v1-provider.js
"use client";
/**
 * TamboV1Provider - Main Provider for v1 API
 *
 * Composes the necessary providers for the v1 SDK:
 * - TamboClientProvider: API client and authentication
 * - TamboRegistryProvider: Component and tool registration
 * - TamboContextHelpersProvider: Context helper functions
 * - TamboV1StreamProvider: Streaming state management
 *
 * This provider should wrap your entire application or the portion
 * that needs access to Tambo v1 functionality.
 */
import React, { createContext, useContext, useState, } from "react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { TamboClientProvider, } from "../../providers/tambo-client-provider";
import { TamboRegistryProvider, } from "../../providers/tambo-registry-provider";
import { TamboContextHelpersProvider } from "../../providers/tambo-context-helpers-provider";
import { TamboV1StreamProvider } from "./tambo-v1-stream-context";
import { TamboV1ThreadInputProvider } from "./tambo-v1-thread-input-provider";
/**
 * Context for v1 SDK configuration.
 * @internal
 */
export const TamboV1ConfigContext = createContext(null);
/**
 * Hook to access v1 SDK configuration.
 * @returns Configuration values including userKey
 * @throws {Error} If used outside TamboV1Provider
 */
export function useTamboV1Config() {
    const config = useContext(TamboV1ConfigContext);
    if (!config) {
        throw new Error("useTamboV1Config must be used within TamboV1Provider");
    }
    return config;
}
/**
 * Main provider for the Tambo v1 SDK.
 *
 * Composes TamboClientProvider, TamboRegistryProvider, and TamboV1StreamProvider
 * to provide a complete context for building AI-powered applications.
 *
 * Threads are managed dynamically through useTamboV1() hook functions:
 * - startNewThread() - Begin a new conversation
 * - switchThread(threadId) - Switch to an existing thread
 * - initThread(threadId) - Initialize a thread for receiving events
 * @param props - Provider configuration
 * @param props.apiKey - Tambo API key for authentication
 * @param props.tamboUrl - Optional custom Tambo API URL
 * @param props.environment - Optional environment configuration
 * @param props.userToken - Optional OAuth token for user authentication
 * @param props.components - Components to register with the AI
 * @param props.tools - Tools to register for client-side execution
 * @param props.mcpServers - MCP servers to register for additional tools/resources
 * @param props.onCallUnregisteredTool - Callback for handling unknown tool calls
 * @param props.resources - Static resources to register with the AI
 * @param props.listResources - Dynamic resource search function (must be paired with getResource)
 * @param props.getResource - Dynamic resource fetch function (must be paired with listResources)
 * @param props.contextHelpers - Configuration for context helper functions
 * @param props.userKey - User key for thread ownership (required if not using userToken)
 * @param props.queryClient - Optional custom React Query client
 * @param props.children - Child components
 * @returns Provider component tree
 * @example
 * ```tsx
 * import { TamboV1Provider } from '@tambo-ai/react/v1';
 *
 * function App() {
 *   return (
 *     <TamboV1Provider
 *       apiKey={process.env.NEXT_PUBLIC_TAMBO_API_KEY!}
 *       components={[WeatherCard, StockChart]}
 *       tools={[searchTool, calculatorTool]}
 *     >
 *       <ChatInterface />
 *     </TamboV1Provider>
 *   );
 * }
 * ```
 */
export function TamboV1Provider({ apiKey, tamboUrl, environment, userToken, components, tools, mcpServers, onCallUnregisteredTool, resources, listResources, getResource, contextHelpers, userKey, queryClient, children, }) {
    // Create a stable default QueryClient if none provided.
    // Using useState to avoid SSR issues with module-level singletons.
    const [defaultQueryClient] = useState(() => new QueryClient({
        defaultOptions: {
            queries: {
                staleTime: 1000,
                retry: 1,
            },
        },
    }));
    const client = queryClient ?? defaultQueryClient;
    // Config is static - created once and never changes
    const config = { userKey };
    return (React.createElement(QueryClientProvider, { client: client },
        React.createElement(TamboClientProvider, { apiKey: apiKey, tamboUrl: tamboUrl, environment: environment, userToken: userToken },
            React.createElement(TamboRegistryProvider, { components: components, tools: tools, mcpServers: mcpServers, onCallUnregisteredTool: onCallUnregisteredTool, resources: resources, listResources: listResources, getResource: getResource },
                React.createElement(TamboContextHelpersProvider, { contextHelpers: contextHelpers },
                    React.createElement(TamboV1ConfigContext.Provider, { value: config },
                        React.createElement(TamboV1StreamProvider, null,
                            React.createElement(TamboV1ThreadInputProvider, null, children))))))));
}
