// @tambo-ai/react/esm/v1/providers/tambo-v1-thread-input-provider.js
"use client";
/**
 * TamboV1ThreadInputProvider - Shared Thread Input Context for v1 API
 *
 * Provides shared input state across all components, enabling features like
 * suggestions to update the input field directly.
 *
 * This mirrors the beta SDK's TamboThreadInputProvider pattern.
 */
import React, { createContext, useCallback, useContext, useState, } from "react";
import { useMessageImages, } from "../../hooks/use-message-images";
import { useTamboMutation, } from "../../hooks/react-query-hooks";
import { useTamboV1SendMessage } from "../hooks/use-tambo-v1-send-message";
import { useStreamState } from "./tambo-v1-stream-context";
// Error messages for various input-related error scenarios.
// TODO: Reintroduce explicit `NETWORK` and `SERVER` keys once `submit()` maps
// failures into a small, stable set of user-facing error codes (at minimum:
// connectivity failures vs non-2xx responses).
export const INPUT_ERROR_MESSAGES = {
    EMPTY: "Message cannot be empty",
    VALIDATION: "Invalid message format",
};
function stagedImageToResourceContent(image) {
    if (!image.dataUrl.startsWith("data:")) {
        throw new Error(INPUT_ERROR_MESSAGES.VALIDATION);
    }
    const commaIndex = image.dataUrl.indexOf(",");
    if (commaIndex === -1) {
        throw new Error(INPUT_ERROR_MESSAGES.VALIDATION);
    }
    const header = image.dataUrl.slice("data:".length, commaIndex);
    const [mimeType, ...params] = header.split(";");
    const isBase64 = params.includes("base64");
    if (mimeType !== image.type || !isBase64) {
        throw new Error(INPUT_ERROR_MESSAGES.VALIDATION);
    }
    return {
        type: "resource",
        resource: {
            name: image.name,
            mimeType: image.type,
            // Shared.Resource.blob expects base64-encoded data; this is the base64
            // payload from the `data:` URL.
            blob: image.dataUrl.slice(commaIndex + 1),
        },
    };
}
/**
 * Context for thread input state.
 * @internal
 */
export const TamboV1ThreadInputContext = createContext(undefined);
/**
 * Provider that manages shared thread input state across all components.
 *
 * This ensures that useTamboV1ThreadInput, useTamboV1Suggestions, and other components
 * all share the same input state.
 * @param props - Provider props
 * @param props.children - Child components
 * @returns Thread input context provider
 */
export function TamboV1ThreadInputProvider({ children }) {
    const [inputValue, setInputValue] = useState("");
    const [threadId, setThreadId] = useState(undefined);
    const imageState = useMessageImages();
    const streamState = useStreamState();
    // Use the current thread from stream state if no explicit threadId is set
    const inheritedThreadId = streamState.currentThreadId ?? undefined;
    const effectiveThreadId = threadId ?? inheritedThreadId;
    const shouldAdoptThreadId = threadId === undefined && inheritedThreadId === undefined;
    const sendMessage = useTamboV1SendMessage(effectiveThreadId);
    const submitFn = useCallback(async (options) => {
        const trimmedValue = inputValue.trim();
        // Check if we have content to send
        if (!trimmedValue && imageState.images.length === 0) {
            throw new Error(INPUT_ERROR_MESSAGES.EMPTY);
        }
        const content = [];
        if (trimmedValue) {
            content.push({ type: "text", text: trimmedValue });
        }
        for (const image of imageState.images) {
            content.push(stagedImageToResourceContent(image));
        }
        const result = await sendMessage.mutateAsync({
            message: {
                role: "user",
                content,
            },
            debug: options?.debug,
        });
        // Clear input and images after successful submission
        setInputValue("");
        imageState.clearImages();
        // Update threadId if a new thread was created
        if (result.threadId && shouldAdoptThreadId) {
            setThreadId(result.threadId);
        }
        return result;
    }, 
    // `stagedImageToResourceContent` is a pure module-level helper (not a hook value).
    [inputValue, imageState, sendMessage, shouldAdoptThreadId]);
    const { mutateAsync: submitAsync, mutate: _unusedSubmit, ...mutationState } = useTamboMutation({
        mutationFn: submitFn,
    });
    const contextValue = {
        ...mutationState,
        value: inputValue,
        setValue: setInputValue,
        submit: submitAsync,
        images: imageState.images,
        addImage: imageState.addImage,
        addImages: imageState.addImages,
        removeImage: imageState.removeImage,
        clearImages: imageState.clearImages,
        threadId: effectiveThreadId,
        setThreadId,
    };
    return (React.createElement(TamboV1ThreadInputContext.Provider, { value: contextValue }, children));
}
/**
 * Hook to access the shared thread input state.
 *
 * All components using this hook share the same input state, enabling
 * features like suggestions to update the input field directly.
 * @returns The shared thread input context
 * @throws {Error} If used outside TamboV1ThreadInputProvider
 * @example
 * ```tsx
 * function ChatInput() {
 *   const { value, setValue, submit, isPending } = useTamboV1ThreadInput();
 *
 *   return (
 *     <form onSubmit={(e) => { e.preventDefault(); submit(); }}>
 *       <input
 *         value={value}
 *         onChange={(e) => setValue(e.target.value)}
 *         disabled={isPending}
 *       />
 *       <button type="submit" disabled={isPending}>Send</button>
 *     </form>
 *   );
 * }
 * ```
 */
export function useTamboV1ThreadInput() {
    const context = useContext(TamboV1ThreadInputContext);
    if (!context) {
        throw new Error("useTamboV1ThreadInput must be used within TamboV1ThreadInputProvider");
    }
    return context;
}
