// @tambo-ai/typescript-sdk/resources/beta/projects/projects.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../../core/resource.mjs";
import * as APIKeyAPI from "./api-key.mjs";
import { APIKey } from "./api-key.mjs";
import * as ProviderKeyAPI from "./provider-key.mjs";
import { ProviderKey, } from "./provider-key.mjs";
import { path } from "../../../internal/utils/path.mjs";
export class Projects extends APIResource {
    constructor() {
        super(...arguments);
        this.apiKey = new APIKeyAPI.APIKey(this._client);
        this.providerKey = new ProviderKeyAPI.ProviderKey(this._client);
    }
    /**
     * @example
     * ```ts
     * const project = await client.beta.projects.retrieve('id');
     * ```
     */
    retrieve(id, options) {
        return this._client.get(path `/projects/${id}`, options);
    }
    /**
     * @example
     * ```ts
     * const project = await client.beta.projects.delete('id');
     * ```
     */
    delete(id, options) {
        return this._client.delete(path `/projects/${id}`, options);
    }
    /**
     * @example
     * ```ts
     * const response = await client.beta.projects.getCurrent();
     * ```
     */
    getCurrent(options) {
        return this._client.get('/projects', options);
    }
}
Projects.APIKey = APIKey;
Projects.ProviderKey = ProviderKey;
