// @tambo-ai/typescript-sdk/resources/beta/projects/projects.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Projects = void 0;
const tslib_1 = require("../../../internal/tslib.js");
const resource_1 = require("../../../core/resource.js");
const APIKeyAPI = tslib_1.__importStar(require("./api-key.js"));
const api_key_1 = require("./api-key.js");
const ProviderKeyAPI = tslib_1.__importStar(require("./provider-key.js"));
const provider_key_1 = require("./provider-key.js");
const path_1 = require("../../../internal/utils/path.js");
class Projects extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.apiKey = new APIKeyAPI.APIKey(this._client);
        this.providerKey = new ProviderKeyAPI.ProviderKey(this._client);
    }
    /**
     * @example
     * ```ts
     * const project = await client.beta.projects.retrieve('id');
     * ```
     */
    retrieve(id, options) {
        return this._client.get((0, path_1.path) `/projects/${id}`, options);
    }
    /**
     * @example
     * ```ts
     * const project = await client.beta.projects.delete('id');
     * ```
     */
    delete(id, options) {
        return this._client.delete((0, path_1.path) `/projects/${id}`, options);
    }
    /**
     * @example
     * ```ts
     * const response = await client.beta.projects.getCurrent();
     * ```
     */
    getCurrent(options) {
        return this._client.get('/projects', options);
    }
}
exports.Projects = Projects;
Projects.APIKey = api_key_1.APIKey;
Projects.ProviderKey = provider_key_1.ProviderKey;
