// @tambo-ai/react/dist/v1/index.js
"use strict";
/**
 * `@tambo-ai/react/v1` - React SDK for Tambo v1 API
 *
 * Provides React hooks and providers for building AI-powered applications
 * using the v1 streaming API with AG-UI protocol.
 *
 * ## Authentication & Thread Ownership
 *
 * All thread operations require user identification. Provide ONE of:
 * - `userKey` - Direct user identifier (for server-side or trusted environments)
 * - `userToken` - OAuth bearer token containing the userKey (for client-side apps)
 *
 * Threads are scoped to the userKey - each user only sees their own threads.
 *
 * ## Quick Start
 *
 * ```tsx
 * import {
 *   TamboV1Provider,
 *   useTamboV1,
 *   useTamboV1ThreadInput,
 * } from '@tambo-ai/react/v1';
 *
 * function App() {
 *   return (
 *     <TamboV1Provider
 *       apiKey={process.env.NEXT_PUBLIC_TAMBO_API_KEY!}
 *       userKey={currentUserId} // Required: identifies thread owner
 *       components={[WeatherCard]}
 *       tools={[searchTool]}
 *     >
 *       <ChatInterface />
 *     </TamboV1Provider>
 *   );
 * }
 *
 * function ChatInterface() {
 *   const [threadId, setThreadId] = useState<string>();
 *   const { messages, isStreaming } = useTamboV1(threadId);
 *   const { value, setValue, submit, isPending } = useTamboV1ThreadInput(threadId);
 *
 *   const handleSubmit = async (e: React.FormEvent) => {
 *     e.preventDefault();
 *     const result = await submit();
 *     if (!threadId) setThreadId(result.threadId);
 *   };
 *
 *   return (
 *     <form onSubmit={handleSubmit}>
 *       {messages.map(msg => <Message key={msg.id} message={msg} />)}
 *       {isStreaming && <LoadingIndicator />}
 *       <input value={value} onChange={(e) => setValue(e.target.value)} />
 *       <button disabled={isPending}>Send</button>
 *     </form>
 *   );
 * }
 * ```
 *
 * ## Type Imports
 *
 * Types are imported directly from specific files:
 * - Thread state: `import type { TamboV1Thread } from '@tambo-ai/react/v1/types/thread'`
 * - Messages: `import type { TamboV1Message } from '@tambo-ai/react/v1/types/message'`
 * - Custom events: `import type { ComponentStartEvent } from '@tambo-ai/react/v1/types/event'`
 *
 * SDK types: `import type { ... } from '@tambo-ai/typescript-sdk'`
 * AG-UI events: `import { EventType, type BaseEvent } from '@ag-ui/core'`
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MCPTransport = exports.currentTimeContextHelper = exports.currentPageContextHelper = exports.defineTool = exports.useTamboVoice = exports.useMessageImages = exports.useTamboClient = exports.useTamboV1Suggestions = exports.useTamboV1ComponentState = exports.useTamboV1ThreadList = exports.useTamboV1Thread = exports.useTamboV1ThreadInput = exports.useTamboV1 = exports.TamboV1StubProvider = exports.TamboV1ThreadInputProvider = exports.useTamboContextHelpers = exports.TamboContextHelpersProvider = exports.useIsTamboTokenUpdating = exports.TamboClientProvider = exports.TamboRegistryProvider = exports.useTamboV1Config = exports.TamboV1Provider = void 0;
// =============================================================================
// Providers
// =============================================================================
var tambo_v1_provider_1 = require("./providers/tambo-v1-provider");
Object.defineProperty(exports, "TamboV1Provider", { enumerable: true, get: function () { return tambo_v1_provider_1.TamboV1Provider; } });
Object.defineProperty(exports, "useTamboV1Config", { enumerable: true, get: function () { return tambo_v1_provider_1.useTamboV1Config; } });
// Re-export registry provider from beta SDK (works with v1)
var tambo_registry_provider_1 = require("../providers/tambo-registry-provider");
Object.defineProperty(exports, "TamboRegistryProvider", { enumerable: true, get: function () { return tambo_registry_provider_1.TamboRegistryProvider; } });
// Re-export client provider from beta SDK (works with v1)
var tambo_client_provider_1 = require("../providers/tambo-client-provider");
Object.defineProperty(exports, "TamboClientProvider", { enumerable: true, get: function () { return tambo_client_provider_1.TamboClientProvider; } });
Object.defineProperty(exports, "useIsTamboTokenUpdating", { enumerable: true, get: function () { return tambo_client_provider_1.useIsTamboTokenUpdating; } });
// Re-export context helpers from beta SDK (works with v1)
var tambo_context_helpers_provider_1 = require("../providers/tambo-context-helpers-provider");
Object.defineProperty(exports, "TamboContextHelpersProvider", { enumerable: true, get: function () { return tambo_context_helpers_provider_1.TamboContextHelpersProvider; } });
Object.defineProperty(exports, "useTamboContextHelpers", { enumerable: true, get: function () { return tambo_context_helpers_provider_1.useTamboContextHelpers; } });
// Export v1 thread input provider for advanced composition
var tambo_v1_thread_input_provider_1 = require("./providers/tambo-v1-thread-input-provider");
Object.defineProperty(exports, "TamboV1ThreadInputProvider", { enumerable: true, get: function () { return tambo_v1_thread_input_provider_1.TamboV1ThreadInputProvider; } });
// Export v1 stub provider for testing
var tambo_v1_stub_provider_1 = require("./providers/tambo-v1-stub-provider");
Object.defineProperty(exports, "TamboV1StubProvider", { enumerable: true, get: function () { return tambo_v1_stub_provider_1.TamboV1StubProvider; } });
// =============================================================================
// Hooks
// =============================================================================
var use_tambo_v1_1 = require("./hooks/use-tambo-v1");
Object.defineProperty(exports, "useTamboV1", { enumerable: true, get: function () { return use_tambo_v1_1.useTamboV1; } });
var use_tambo_v1_thread_input_1 = require("./hooks/use-tambo-v1-thread-input");
Object.defineProperty(exports, "useTamboV1ThreadInput", { enumerable: true, get: function () { return use_tambo_v1_thread_input_1.useTamboV1ThreadInput; } });
var use_tambo_v1_thread_1 = require("./hooks/use-tambo-v1-thread");
Object.defineProperty(exports, "useTamboV1Thread", { enumerable: true, get: function () { return use_tambo_v1_thread_1.useTamboV1Thread; } });
var use_tambo_v1_thread_list_1 = require("./hooks/use-tambo-v1-thread-list");
Object.defineProperty(exports, "useTamboV1ThreadList", { enumerable: true, get: function () { return use_tambo_v1_thread_list_1.useTamboV1ThreadList; } });
var use_tambo_v1_component_state_1 = require("./hooks/use-tambo-v1-component-state");
Object.defineProperty(exports, "useTamboV1ComponentState", { enumerable: true, get: function () { return use_tambo_v1_component_state_1.useTamboV1ComponentState; } });
var use_tambo_v1_suggestions_1 = require("./hooks/use-tambo-v1-suggestions");
Object.defineProperty(exports, "useTamboV1Suggestions", { enumerable: true, get: function () { return use_tambo_v1_suggestions_1.useTamboV1Suggestions; } });
// Re-export client hook from beta SDK (works with v1)
var tambo_client_provider_2 = require("../providers/tambo-client-provider");
Object.defineProperty(exports, "useTamboClient", { enumerable: true, get: function () { return tambo_client_provider_2.useTamboClient; } });
// Re-export image handling hook (pure React, API-agnostic)
var use_message_images_1 = require("../hooks/use-message-images");
Object.defineProperty(exports, "useMessageImages", { enumerable: true, get: function () { return use_message_images_1.useMessageImages; } });
// Re-export voice input hook (uses standalone transcription API)
var use_tambo_voice_1 = require("../hooks/use-tambo-voice");
Object.defineProperty(exports, "useTamboVoice", { enumerable: true, get: function () { return use_tambo_voice_1.useTamboVoice; } });
// =============================================================================
// Re-exports from Beta SDK (compatible with v1)
// =============================================================================
// Tool definition helper
var registry_1 = require("../util/registry");
Object.defineProperty(exports, "defineTool", { enumerable: true, get: function () { return registry_1.defineTool; } });
// Built-in context helpers
var context_helpers_1 = require("../context-helpers");
Object.defineProperty(exports, "currentPageContextHelper", { enumerable: true, get: function () { return context_helpers_1.currentPageContextHelper; } });
Object.defineProperty(exports, "currentTimeContextHelper", { enumerable: true, get: function () { return context_helpers_1.currentTimeContextHelper; } });
// MCP server types
var mcp_server_info_1 = require("../model/mcp-server-info");
Object.defineProperty(exports, "MCPTransport", { enumerable: true, get: function () { return mcp_server_info_1.MCPTransport; } });
