// @tambo-ai/typescript-sdk/resources/beta/beta.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Beta = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const AudioAPI = tslib_1.__importStar(require("./audio.js"));
const audio_1 = require("./audio.js");
const AuthAPI = tslib_1.__importStar(require("./auth.js"));
const auth_1 = require("./auth.js");
const RegistryAPI = tslib_1.__importStar(require("./registry.js"));
const registry_1 = require("./registry.js");
const ProjectsAPI = tslib_1.__importStar(require("./projects/projects.js"));
const projects_1 = require("./projects/projects.js");
const ThreadsAPI = tslib_1.__importStar(require("./threads/threads.js"));
const threads_1 = require("./threads/threads.js");
class Beta extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.registry = new RegistryAPI.Registry(this._client);
        this.projects = new ProjectsAPI.Projects(this._client);
        this.threads = new ThreadsAPI.Threads(this._client);
        this.auth = new AuthAPI.Auth(this._client);
        this.audio = new AudioAPI.Audio(this._client);
    }
}
exports.Beta = Beta;
Beta.Registry = registry_1.Registry;
Beta.Projects = projects_1.Projects;
Beta.Threads = threads_1.Threads;
Beta.Auth = auth_1.Auth;
Beta.Audio = audio_1.Audio;
