// @tambo-ai/typescript-sdk/resources/beta/beta.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../core/resource.mjs";
import * as AudioAPI from "./audio.mjs";
import { Audio } from "./audio.mjs";
import * as AuthAPI from "./auth.mjs";
import { Auth, } from "./auth.mjs";
import * as RegistryAPI from "./registry.mjs";
import { Registry } from "./registry.mjs";
import * as ProjectsAPI from "./projects/projects.mjs";
import { Projects, } from "./projects/projects.mjs";
import * as ThreadsAPI from "./threads/threads.mjs";
import { Threads, } from "./threads/threads.mjs";
export class Beta extends APIResource {
    constructor() {
        super(...arguments);
        this.registry = new RegistryAPI.Registry(this._client);
        this.projects = new ProjectsAPI.Projects(this._client);
        this.threads = new ThreadsAPI.Threads(this._client);
        this.auth = new AuthAPI.Auth(this._client);
        this.audio = new AudioAPI.Audio(this._client);
    }
}
Beta.Registry = Registry;
Beta.Projects = Projects;
Beta.Threads = Threads;
Beta.Auth = Auth;
Beta.Audio = Audio;
