// @tambo-ai/typescript-sdk/resources/beta/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Threads = exports.Registry = exports.Projects = exports.Beta = exports.Auth = exports.Audio = void 0;
var audio_1 = require("./audio.js");
Object.defineProperty(exports, "Audio", { enumerable: true, get: function () { return audio_1.Audio; } });
var auth_1 = require("./auth.js");
Object.defineProperty(exports, "Auth", { enumerable: true, get: function () { return auth_1.Auth; } });
var beta_1 = require("./beta.js");
Object.defineProperty(exports, "Beta", { enumerable: true, get: function () { return beta_1.Beta; } });
var index_1 = require("./projects/index.js");
Object.defineProperty(exports, "Projects", { enumerable: true, get: function () { return index_1.Projects; } });
var registry_1 = require("./registry.js");
Object.defineProperty(exports, "Registry", { enumerable: true, get: function () { return registry_1.Registry; } });
var index_2 = require("./threads/index.js");
Object.defineProperty(exports, "Threads", { enumerable: true, get: function () { return index_2.Threads; } });
