// @tambo-ai/react/dist/v1/providers/tambo-v1-thread-input-provider.js
"use strict";
"use client";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TamboV1ThreadInputContext = exports.INPUT_ERROR_MESSAGES = void 0;
exports.TamboV1ThreadInputProvider = TamboV1ThreadInputProvider;
exports.useTamboV1ThreadInput = useTamboV1ThreadInput;
/**
 * TamboV1ThreadInputProvider - Shared Thread Input Context for v1 API
 *
 * Provides shared input state across all components, enabling features like
 * suggestions to update the input field directly.
 *
 * This mirrors the beta SDK's TamboThreadInputProvider pattern.
 */
const react_1 = __importStar(require("react"));
const use_message_images_1 = require("../../hooks/use-message-images");
const react_query_hooks_1 = require("../../hooks/react-query-hooks");
const use_tambo_v1_send_message_1 = require("../hooks/use-tambo-v1-send-message");
const tambo_v1_stream_context_1 = require("./tambo-v1-stream-context");
// Error messages for various input-related error scenarios.
// TODO: Reintroduce explicit `NETWORK` and `SERVER` keys once `submit()` maps
// failures into a small, stable set of user-facing error codes (at minimum:
// connectivity failures vs non-2xx responses).
exports.INPUT_ERROR_MESSAGES = {
    EMPTY: "Message cannot be empty",
    VALIDATION: "Invalid message format",
};
function stagedImageToResourceContent(image) {
    if (!image.dataUrl.startsWith("data:")) {
        throw new Error(exports.INPUT_ERROR_MESSAGES.VALIDATION);
    }
    const commaIndex = image.dataUrl.indexOf(",");
    if (commaIndex === -1) {
        throw new Error(exports.INPUT_ERROR_MESSAGES.VALIDATION);
    }
    const header = image.dataUrl.slice("data:".length, commaIndex);
    const [mimeType, ...params] = header.split(";");
    const isBase64 = params.includes("base64");
    if (mimeType !== image.type || !isBase64) {
        throw new Error(exports.INPUT_ERROR_MESSAGES.VALIDATION);
    }
    return {
        type: "resource",
        resource: {
            name: image.name,
            mimeType: image.type,
            // Shared.Resource.blob expects base64-encoded data; this is the base64
            // payload from the `data:` URL.
            blob: image.dataUrl.slice(commaIndex + 1),
        },
    };
}
/**
 * Context for thread input state.
 * @internal
 */
exports.TamboV1ThreadInputContext = (0, react_1.createContext)(undefined);
/**
 * Provider that manages shared thread input state across all components.
 *
 * This ensures that useTamboV1ThreadInput, useTamboV1Suggestions, and other components
 * all share the same input state.
 * @param props - Provider props
 * @param props.children - Child components
 * @returns Thread input context provider
 */
function TamboV1ThreadInputProvider({ children }) {
    const [inputValue, setInputValue] = (0, react_1.useState)("");
    const [threadId, setThreadId] = (0, react_1.useState)(undefined);
    const imageState = (0, use_message_images_1.useMessageImages)();
    const streamState = (0, tambo_v1_stream_context_1.useStreamState)();
    // Use the current thread from stream state if no explicit threadId is set
    const inheritedThreadId = streamState.currentThreadId ?? undefined;
    const effectiveThreadId = threadId ?? inheritedThreadId;
    const shouldAdoptThreadId = threadId === undefined && inheritedThreadId === undefined;
    const sendMessage = (0, use_tambo_v1_send_message_1.useTamboV1SendMessage)(effectiveThreadId);
    const submitFn = (0, react_1.useCallback)(async (options) => {
        const trimmedValue = inputValue.trim();
        // Check if we have content to send
        if (!trimmedValue && imageState.images.length === 0) {
            throw new Error(exports.INPUT_ERROR_MESSAGES.EMPTY);
        }
        const content = [];
        if (trimmedValue) {
            content.push({ type: "text", text: trimmedValue });
        }
        for (const image of imageState.images) {
            content.push(stagedImageToResourceContent(image));
        }
        const result = await sendMessage.mutateAsync({
            message: {
                role: "user",
                content,
            },
            debug: options?.debug,
        });
        // Clear input and images after successful submission
        setInputValue("");
        imageState.clearImages();
        // Update threadId if a new thread was created
        if (result.threadId && shouldAdoptThreadId) {
            setThreadId(result.threadId);
        }
        return result;
    }, 
    // `stagedImageToResourceContent` is a pure module-level helper (not a hook value).
    [inputValue, imageState, sendMessage, shouldAdoptThreadId]);
    const { mutateAsync: submitAsync, mutate: _unusedSubmit, ...mutationState } = (0, react_query_hooks_1.useTamboMutation)({
        mutationFn: submitFn,
    });
    const contextValue = {
        ...mutationState,
        value: inputValue,
        setValue: setInputValue,
        submit: submitAsync,
        images: imageState.images,
        addImage: imageState.addImage,
        addImages: imageState.addImages,
        removeImage: imageState.removeImage,
        clearImages: imageState.clearImages,
        threadId: effectiveThreadId,
        setThreadId,
    };
    return (react_1.default.createElement(exports.TamboV1ThreadInputContext.Provider, { value: contextValue }, children));
}
/**
 * Hook to access the shared thread input state.
 *
 * All components using this hook share the same input state, enabling
 * features like suggestions to update the input field directly.
 * @returns The shared thread input context
 * @throws {Error} If used outside TamboV1ThreadInputProvider
 * @example
 * ```tsx
 * function ChatInput() {
 *   const { value, setValue, submit, isPending } = useTamboV1ThreadInput();
 *
 *   return (
 *     <form onSubmit={(e) => { e.preventDefault(); submit(); }}>
 *       <input
 *         value={value}
 *         onChange={(e) => setValue(e.target.value)}
 *         disabled={isPending}
 *       />
 *       <button type="submit" disabled={isPending}>Send</button>
 *     </form>
 *   );
 * }
 * ```
 */
function useTamboV1ThreadInput() {
    const context = (0, react_1.useContext)(exports.TamboV1ThreadInputContext);
    if (!context) {
        throw new Error("useTamboV1ThreadInput must be used within TamboV1ThreadInputProvider");
    }
    return context;
}
