// @tambo-ai/react/dist/providers/tambo-client-provider.js
"use strict";
"use client";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useIsTamboTokenUpdating = exports.useTamboQueryClient = exports.useTamboClient = exports.TamboClientProvider = exports.TamboClientContext = void 0;
const typescript_sdk_1 = __importDefault(require("@tambo-ai/typescript-sdk"));
const react_query_1 = require("@tanstack/react-query");
const react_1 = __importStar(require("react"));
const package_json_1 = __importDefault(require("../../package.json"));
const use_tambo_session_token_1 = require("./hooks/use-tambo-session-token");
exports.TamboClientContext = (0, react_1.createContext)(undefined);
/**
 * The TamboClientProvider is a React provider that provides a TamboAI client
 * and a query client to the descendants of the provider.
 * @param props - The props for the TamboClientProvider
 * @param props.children - The children to wrap
 * @param props.tamboUrl - The URL of the Tambo API
 * @param props.apiKey - The API key for the Tambo API
 * @param props.environment - The environment to use for the Tambo API
 * @param props.userToken - The oauth access token to use to identify the user in the Tambo API
 * @param props.additionalHeaders - Additional headers to include in all requests
 * @returns The TamboClientProvider component
 */
const TamboClientProvider = ({ children, tamboUrl, apiKey, environment, userToken, additionalHeaders, }) => {
    const tamboConfig = (0, react_1.useMemo)(() => ({
        apiKey,
        defaultHeaders: {
            "X-Tambo-React-Version": package_json_1.default.version,
            ...additionalHeaders,
        },
        baseURL: tamboUrl ?? undefined,
        environment: environment ?? undefined,
    }), [additionalHeaders, apiKey, tamboUrl, environment]);
    const client = (0, react_1.useMemo)(() => new typescript_sdk_1.default(tamboConfig), [tamboConfig]);
    const queryClient = (0, react_1.useMemo)(() => new react_query_1.QueryClient(), []);
    // Keep the session token updated and get the updating state
    const { isFetching: isUpdatingToken } = (0, use_tambo_session_token_1.useTamboSessionToken)(client, queryClient, userToken);
    return (react_1.default.createElement(exports.TamboClientContext.Provider, { value: {
            client,
            queryClient,
            isUpdatingToken,
            additionalHeaders: tamboConfig.defaultHeaders,
        } }, children));
};
exports.TamboClientProvider = TamboClientProvider;
/**
 * The useTamboClient hook provides access to the TamboAI client
 * to the descendants of the TamboClientProvider.
 * @returns The TamboAI client
 */
const useTamboClient = () => {
    const context = react_1.default.useContext(exports.TamboClientContext);
    if (context === undefined) {
        throw new Error("useTamboClient must be used within a TamboClientProvider");
    }
    return context.client;
};
exports.useTamboClient = useTamboClient;
/**
 * The useTamboQueryClient hook provides access to the tambo-specific query client
 * to the descendants of the TamboClientProvider.
 * @returns The tambo-specific query client
 * @private
 */
const useTamboQueryClient = () => {
    const context = react_1.default.useContext(exports.TamboClientContext);
    if (context === undefined) {
        throw new Error("useTamboQueryClient must be used within a TamboClientProvider");
    }
    return context.queryClient;
};
exports.useTamboQueryClient = useTamboQueryClient;
/**
 * Hook to check if the session token is currently being updated
 * @returns true if the token is being refreshed, false otherwise
 */
const useIsTamboTokenUpdating = () => {
    const context = react_1.default.useContext(exports.TamboClientContext);
    if (context === undefined) {
        throw new Error("useIsTamboTokenUpdating must be used within a TamboClientProvider");
    }
    return context.isUpdatingToken;
};
exports.useIsTamboTokenUpdating = useIsTamboTokenUpdating;
