// @tambo-ai/react/dist/providers/tambo-context-attachment-provider.js
"use strict";
"use client";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TamboContextAttachmentProvider = TamboContextAttachmentProvider;
exports.useTamboContextAttachment = useTamboContextAttachment;
const react_1 = __importStar(require("react"));
const tambo_context_helpers_provider_1 = require("./tambo-context-helpers-provider");
const ContextAttachmentContext = (0, react_1.createContext)(null);
const CONTEXT_ATTACHMENTS_HELPER_KEY = "contextAttachments";
/**
 * Provider that manages context attachments for the next user message.
 * - **Included by default** in TamboProvider - no need to wrap separately
 * - Use `useTamboContextAttachment()` hook to manage context attachments
 * **What it does:**
 * - Stores context attachments that will be sent with the next message
 * - Automatically registers/deregisters context helpers for each attachment
 * - Context helpers are automatically collected during message submission
 * - Context attachments are cleared after message submission (one-time use)
 *
 * **Note:** Context attachments are automatically included in additionalContext when
 * the next message is sent. They are cleared after submission.
 * @param props - The props for the TamboContextAttachmentProvider
 * @param props.children - The children to wrap
 * @returns The TamboContextAttachmentProvider component
 */
function TamboContextAttachmentProvider({ children, }) {
    const { addContextHelper, removeContextHelper } = (0, tambo_context_helpers_provider_1.useTamboContextHelpers)();
    const [attachments, setAttachments] = (0, react_1.useState)([]);
    (0, react_1.useEffect)(() => {
        addContextHelper(CONTEXT_ATTACHMENTS_HELPER_KEY, () => {
            if (attachments.length === 0) {
                return null;
            }
            return attachments.map((attachment) => ({
                id: attachment.id,
                displayName: attachment.displayName,
                context: attachment.context,
                type: attachment.type,
            }));
        });
        return () => {
            removeContextHelper(CONTEXT_ATTACHMENTS_HELPER_KEY);
        };
    }, [attachments, addContextHelper, removeContextHelper]);
    /**
     * Adds a new context attachment that will be included with the next user message.
     * The attachment is automatically registered as part of the merged context helper.
     * @param contextAttachment - The context attachment input (context, optional displayName, optional type)
     * @returns The created ContextAttachment object with a unique ID
     * @example
     * ```tsx
     * const attachment = addContextAttachment({
     *   context: "The contents of File.txt",
     *   displayName: "File.txt",
     *   type: "file"
     * });
     * ```
     */
    const addContextAttachment = (0, react_1.useCallback)((contextAttachment) => {
        const id = crypto.randomUUID();
        const attachment = {
            id,
            displayName: contextAttachment.displayName,
            context: contextAttachment.context,
            type: contextAttachment.type,
        };
        setAttachments((prev) => [...prev, attachment]);
        return attachment;
    }, []);
    /**
     * Removes a context attachment by its ID.
     */
    const removeContextAttachment = (0, react_1.useCallback)((id) => {
        setAttachments((prev) => prev.filter((c) => c.id !== id));
    }, []);
    /**
     * Removes all context attachments at once.
     */
    const clearContextAttachments = (0, react_1.useCallback)(() => {
        setAttachments([]);
    }, []);
    const value = (0, react_1.useMemo)(() => ({
        attachments,
        addContextAttachment,
        removeContextAttachment,
        clearContextAttachments,
    }), [
        attachments,
        addContextAttachment,
        removeContextAttachment,
        clearContextAttachments,
    ]);
    return (react_1.default.createElement(ContextAttachmentContext.Provider, { value: value }, children));
}
/**
 * Hook to access context attachment state and methods.
 * **Must be used within a `TamboProvider`** - throws an error otherwise.
 * @throws {Error} If used outside of TamboProvider
 * @returns The context attachment state and methods
 * @example
 * ```tsx
 * const { addContextAttachment, attachments, clearContextAttachments } = useTamboContextAttachment();
 *
 * // Add a context attachment for the next message
 * const attachment = addContextAttachment({
 *   context: "The contents of File.txt",
 *   displayName: "File.txt", // optional
 *   type: "file" // optional
 * });
 *
 * // Remove a context attachment
 * removeContextAttachment(attachment.id);
 *
 * // Clear all context attachments
 * clearContextAttachments();
 * ```
 */
function useTamboContextAttachment() {
    const context = (0, react_1.useContext)(ContextAttachmentContext);
    if (!context) {
        throw new Error("useTamboContextAttachment must be used within a TamboContextAttachmentProvider");
    }
    return context;
}
