// @tambo-ai/react/dist/providers/tambo-component-provider.js
"use strict";
"use client";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTamboComponent = exports.TamboComponentProvider = void 0;
const react_1 = __importStar(require("react"));
const tambo_client_provider_1 = require("./tambo-client-provider");
const tambo_registry_provider_1 = require("./tambo-registry-provider");
const TamboComponentContext = (0, react_1.createContext)({
    /**
     *
     */
    registerComponent: () => { },
    /**
     *
     */
    registerTool: () => { },
    /**
     *
     */
    registerTools: () => { },
    /**
     *
     */
    addToolAssociation: () => { },
});
/**
 * The TamboComponentProvider is a React provider that provides component
 * registration services to the descendants of the provider.
 * @param props - The props for the TamboComponentProvider
 * @param props.children - The children to wrap
 * @returns The TamboComponentProvider component
 */
const TamboComponentProvider = ({ children, }) => {
    const client = (0, tambo_client_provider_1.useTamboClient)();
    const { registerComponent, addToolAssociation, registerTool, registerTools } = (0, tambo_registry_provider_1.useTamboRegistry)();
    const value = {
        client,
        registerComponent,
        registerTool,
        registerTools,
        addToolAssociation,
    };
    return (react_1.default.createElement(TamboComponentContext.Provider, { value: value }, children));
};
exports.TamboComponentProvider = TamboComponentProvider;
/**
 * The useTamboComponent hook provides access to the component registration
 * services to the descendants of the TamboComponentProvider.
 * @returns The component registration services
 */
const useTamboComponent = () => {
    return (0, react_1.useContext)(TamboComponentContext);
};
exports.useTamboComponent = useTamboComponent;
