// @tambo-ai/react/dist/providers/tambo-stubs.js
"use strict";
"use client";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TamboStubProvider = void 0;
const react_query_1 = require("@tanstack/react-query");
const react_1 = __importStar(require("react"));
const generate_component_response_1 = require("../model/generate-component-response");
const tambo_client_provider_1 = require("./tambo-client-provider");
const tambo_component_provider_1 = require("./tambo-component-provider");
const tambo_context_attachment_provider_1 = require("./tambo-context-attachment-provider");
const tambo_context_helpers_provider_1 = require("./tambo-context-helpers-provider");
const tambo_interactable_provider_1 = require("./tambo-interactable-provider");
const tambo_provider_1 = require("./tambo-provider");
const tambo_registry_provider_1 = require("./tambo-registry-provider");
const tambo_thread_provider_1 = require("./tambo-thread-provider");
/**
 * Stub client provider that accepts a client and queryClient as props
 * @returns The TamboStubClientProvider component
 */
const TamboStubClientProvider = ({ children, client, queryClient, threads, projectId, contextKey, isUpdatingToken, }) => {
    // Prepopulate the query cache with threads data if provided
    (0, react_1.useEffect)(() => {
        if (threads) {
            // Set the project ID in the cache
            queryClient.setQueryData(["projectId"], projectId);
            // Set the threads data in the cache using the same query key pattern as useTamboThreadList
            queryClient.setQueryData(["threads", projectId, contextKey], threads);
        }
    }, [threads, projectId, contextKey, queryClient]);
    return (react_1.default.createElement(tambo_client_provider_1.TamboClientContext.Provider, { value: {
            client,
            queryClient,
            isUpdatingToken,
        } }, children));
};
/**
 * Stub registry provider that accepts componentList, toolRegistry, and componentToolAssociations as props
 * @returns The TamboStubRegistryProvider component
 */
const TamboStubRegistryProvider = ({ children, componentList, toolRegistry, componentToolAssociations, registerComponent, registerTool, registerTools, addToolAssociation, }) => {
    return (react_1.default.createElement(tambo_registry_provider_1.TamboRegistryContext.Provider, { value: {
            componentList,
            toolRegistry,
            componentToolAssociations,
            mcpServerInfos: [],
            resources: [],
            resourceSource: null,
            registerComponent,
            registerTool,
            registerTools,
            addToolAssociation,
            registerMcpServer: () => { },
            registerMcpServers: () => { },
            registerResource: () => { },
            registerResources: () => { },
            registerResourceSource: () => { },
        } }, children));
};
/**
 * Stub thread provider that accepts all thread context props
 * @returns The TamboStubThreadProvider component
 */
const TamboStubThreadProvider = ({ children, ...threadContextProps }) => {
    // Extract generation stage info from the thread
    const generationStage = threadContextProps.thread?.generationStage ??
        generate_component_response_1.GenerationStage.IDLE;
    const statusMessage = threadContextProps.thread?.statusMessage ?? "";
    return (react_1.default.createElement(tambo_thread_provider_1.TamboThreadContext.Provider, { value: threadContextProps },
        react_1.default.createElement(tambo_thread_provider_1.TamboGenerationStageProvider, { generationStage: generationStage, statusMessage: statusMessage }, children)));
};
/**
 * Default no-op functions that throw errors - used when callbacks are not provided
 * @returns The default callbacks
 */
const createDefaultCallbacks = () => ({
    switchCurrentThread: () => {
        throw new Error("switchCurrentThread not implemented in stub");
    },
    startNewThread: () => {
        throw new Error("startNewThread not implemented in stub");
    },
    updateThreadName: () => {
        throw new Error("updateThreadName not implemented in stub");
    },
    generateThreadName: () => {
        throw new Error("generateThreadName not implemented in stub");
    },
    addThreadMessage: () => {
        throw new Error("addThreadMessage not implemented in stub");
    },
    updateThreadMessage: () => {
        throw new Error("updateThreadMessage not implemented in stub");
    },
    setInputValue: () => {
        throw new Error("setInputValue not implemented in stub");
    },
    sendThreadMessage: () => {
        throw new Error("sendThreadMessage not implemented in stub");
    },
    registerComponent: () => {
        throw new Error("registerComponent not implemented in stub");
    },
    registerTool: () => {
        throw new Error("registerTool not implemented in stub");
    },
    registerTools: () => {
        throw new Error("registerTools not implemented in stub");
    },
    addToolAssociation: () => {
        throw new Error("addToolAssociation not implemented in stub");
    },
    cancel: () => {
        throw new Error("cancel not implemented in stub");
    },
});
/**
 * The TamboStubProvider provides a way to use the Tambo API with static/stub data.
 * This is useful for testing, samples, and static UI development.
 *
 * Unlike TamboProvider, all data must be provided as props rather than being inferred.
 * @example
 * ```tsx
 * import { TamboStubProvider } from '@tambo-ai/react-sdk';
 *
 * const exampleThread = {
 *   id: "example-thread",
 *   messages: [
 *     {
 *       id: "msg-1",
 *       role: "user",
 *       content: [{ type: "text", text: "Hello!" }],
 *       createdAt: new Date().toISOString(),
 *       threadId: "example-thread",
 *       componentState: {},
 *     },
 *     {
 *       id: "msg-2",
 *       role: "assistant",
 *       content: [{ type: "text", text: "Hi there!" }],
 *       createdAt: new Date().toISOString(),
 *       threadId: "example-thread",
 *       componentState: {},
 *     }
 *   ],
 *   createdAt: new Date().toISOString(),
 *   projectId: "example-project",
 *   updatedAt: new Date().toISOString(),
 *   metadata: {},
 * };
 *
 * const exampleThreadList = [exampleThread];
 *
 * function MyComponent() {
 *   return (
 *     <TamboStubProvider
 *       thread={exampleThread}
 *       components={[]}
 *       tools={[]}
 *       threads={exampleThreadList}
 *       projectId="example-project"
 *     >
 *       <MessageThreadFull />
 *     </TamboStubProvider>
 *   );
 * }
 * ```
 * @param props - The props for the TamboStubProvider
 * @param props.children - The children to wrap
 * @param props.thread - The thread data to display
 * @param props.components - Optional components registry
 * @param props.tools - Optional tools registry
 * @param props.threads - Optional threads data to populate thread list (overrides useTamboThreadList)
 * @param props.projectId - Optional project ID for query cache (defaults to thread.projectId)
 * @param props.contextKey - Optional context key for thread list queries
 * @param props.contextHelpers - Optional configuration for which context helpers are enabled/disabled
 * @returns The TamboStubProvider component
 */
const TamboStubProvider = ({ children, thread, components = [], tools = [], threads, projectId, contextKey, contextHelpers, ...overrides }) => {
    const defaults = createDefaultCallbacks();
    // Create stub client and queryClient
    const stubClient = {};
    const stubQueryClient = new react_query_1.QueryClient();
    // Use provided projectId or fall back to thread's projectId
    const resolvedProjectId = projectId ?? thread.projectId;
    // Build component registry from components prop
    const componentList = components.reduce((acc, component) => {
        acc[component.name] = {
            component: component.component,
            loadingComponent: component.loadingComponent,
            name: component.name,
            description: component.description,
            props: component.propsDefinition ?? {},
            contextTools: [],
        };
        return acc;
    }, {});
    // Build tool registry from tools prop
    const toolRegistry = tools.reduce((acc, tool) => {
        acc[tool.name] = tool;
        return acc;
    }, {});
    // Build tool associations from components
    const componentToolAssociations = components.reduce((acc, component) => {
        if (component.associatedTools) {
            acc[component.name] = component.associatedTools.map((tool) => tool.name);
        }
        return acc;
    }, {});
    // Merge defaults with provided overrides
    const threadContextProps = {
        thread,
        currentThreadId: thread.id,
        currentThread: thread,
        threadMap: { [thread.id]: thread },
        setThreadMap: () => { },
        switchCurrentThread: overrides.switchCurrentThread ?? defaults.switchCurrentThread,
        startNewThread: overrides.startNewThread ?? defaults.startNewThread,
        updateThreadName: overrides.updateThreadName ?? defaults.updateThreadName,
        generateThreadName: overrides.generateThreadName ?? defaults.generateThreadName,
        addThreadMessage: overrides.addThreadMessage ?? defaults.addThreadMessage,
        updateThreadMessage: overrides.updateThreadMessage ?? defaults.updateThreadMessage,
        streaming: overrides.streaming ?? true,
        sendThreadMessage: overrides.sendThreadMessage ?? defaults.sendThreadMessage,
        cancel: overrides.cancel ?? defaults.cancel,
        contextKey,
    };
    const componentContextProps = {
        registerComponent: overrides.registerComponent ?? defaults.registerComponent,
        registerTool: overrides.registerTool ?? defaults.registerTool,
        registerTools: overrides.registerTools ?? defaults.registerTools,
        addToolAssociation: overrides.addToolAssociation ?? defaults.addToolAssociation,
    };
    return (react_1.default.createElement(TamboStubClientProvider, { client: stubClient, queryClient: stubQueryClient, threads: threads, projectId: resolvedProjectId, contextKey: contextKey, isUpdatingToken: false },
        react_1.default.createElement(TamboStubRegistryProvider, { componentList: componentList, toolRegistry: toolRegistry, componentToolAssociations: componentToolAssociations, registerComponent: componentContextProps.registerComponent, registerTool: componentContextProps.registerTool, registerTools: componentContextProps.registerTools, addToolAssociation: componentContextProps.addToolAssociation },
            react_1.default.createElement(TamboStubThreadProvider, { ...threadContextProps },
                react_1.default.createElement(tambo_context_helpers_provider_1.TamboContextHelpersProvider, { contextHelpers: contextHelpers },
                    react_1.default.createElement(tambo_context_attachment_provider_1.TamboContextAttachmentProvider, null,
                        react_1.default.createElement(tambo_component_provider_1.TamboComponentProvider, null,
                            react_1.default.createElement(tambo_interactable_provider_1.TamboInteractableProvider, null,
                                react_1.default.createElement(tambo_provider_1.TamboCompositeProvider, null, children)))))))));
};
exports.TamboStubProvider = TamboStubProvider;
