// @tambo-ai/react/dist/providers/tambo-thread-input-provider.js
"use strict";
"use client";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTamboThreadInput = exports.TamboThreadInputProvider = exports.TamboThreadInputContext = exports.INPUT_ERROR_MESSAGES = void 0;
const react_1 = __importStar(require("react"));
const react_query_hooks_1 = require("../hooks/react-query-hooks");
const use_message_images_1 = require("../hooks/use-message-images");
const tambo_mcp_provider_1 = require("../mcp/tambo-mcp-provider");
const thread_input_error_1 = require("../model/thread-input-error");
const validate_input_1 = require("../model/validate-input");
const message_builder_1 = require("../util/message-builder");
const resource_content_resolver_1 = require("../util/resource-content-resolver");
const tambo_interactable_provider_1 = require("./tambo-interactable-provider");
const tambo_registry_provider_1 = require("./tambo-registry-provider");
const tambo_thread_provider_1 = require("./tambo-thread-provider");
/**
 * Error messages for various input-related error scenarios
 * These messages are used to provide user-friendly error feedback
 * @readonly
 */
exports.INPUT_ERROR_MESSAGES = {
    EMPTY: "Message cannot be empty",
    NETWORK: "Network error. Please check your connection",
    SERVER: "Server error. Please try again",
    VALIDATION: "Invalid message format",
};
exports.TamboThreadInputContext = (0, react_1.createContext)(undefined);
/**
 * Provider that manages shared thread input state across all components
 * This ensures that useTamboThreadInput, useTamboSuggestions, and components
 * all share the same input state
 * @param props - The props for the TamboThreadInputProvider
 * @param props.children - The children to render.
 * @returns The thread input context
 */
const TamboThreadInputProvider = ({ children, }) => {
    const { thread, sendThreadMessage, contextKey } = (0, tambo_thread_provider_1.useTamboThread)();
    const [inputValue, setInputValue] = (0, react_1.useState)("");
    const imageState = (0, use_message_images_1.useMessageImages)();
    const mcpServers = (0, tambo_mcp_provider_1.useTamboMcpServers)();
    const { resourceSource } = (0, tambo_registry_provider_1.useTamboRegistry)();
    const { clearInteractableSelections } = (0, tambo_interactable_provider_1.useTamboInteractable)();
    const submit = (0, react_1.useCallback)(async ({ streamResponse, forceToolChoice, additionalContext, resourceNames = {}, } = {}) => {
        // Validate text input if present
        if (inputValue?.trim()) {
            const validation = (0, validate_input_1.validateInput)(inputValue);
            if (!validation.isValid) {
                throw new thread_input_error_1.ThreadInputError(`Cannot submit message: ${validation.error ?? exports.INPUT_ERROR_MESSAGES.VALIDATION}`, { cause: validation.error });
            }
        }
        // Check if we have content to send
        if (!inputValue.trim() && imageState.images.length === 0) {
            throw new thread_input_error_1.ThreadInputError(exports.INPUT_ERROR_MESSAGES.EMPTY, {
                cause: "No text or images to send",
            });
        }
        // Extract resource URIs from the input text and resolve content for client-side resources
        // (registry and client-side MCP servers). Internal Tambo server resources are skipped
        // since the backend can resolve them.
        const resourceUris = (0, resource_content_resolver_1.extractResourceUris)(inputValue);
        const resolvedContent = await (0, resource_content_resolver_1.resolveResourceContents)(resourceUris, mcpServers, resourceSource ?? undefined);
        // Build message content with text, images, resource names, and resolved content
        const messageContent = (0, message_builder_1.buildMessageContent)(inputValue, imageState.images, resourceNames, resolvedContent);
        try {
            await sendThreadMessage(inputValue || "Image message", {
                threadId: thread.id,
                contextKey,
                streamResponse: streamResponse,
                forceToolChoice: forceToolChoice,
                additionalContext: additionalContext,
                content: messageContent,
            });
            clearInteractableSelections();
        }
        catch (error) {
            // Handle image-related errors with friendly messages
            if (imageState.images.length > 0) {
                const errorMessage = error instanceof Error ? error.message.toLowerCase() : "";
                // Backend not yet supporting image content type
                if (errorMessage.includes("unknown content part type: image")) {
                    throw new thread_input_error_1.ThreadInputError("Image attachments are not yet supported by the backend. This feature is coming soon.", { cause: error });
                }
                // Handle specific model vision support errors
                // OpenAI errors
                if (errorMessage.includes("does not support image message content types") ||
                    (errorMessage.includes("invalid model") &&
                        errorMessage.includes("image_url is only supported by certain models"))) {
                    throw new thread_input_error_1.ThreadInputError("This model doesn't support images. Please use GPT-4o, GPT-4 Turbo, or other vision-capable models.", { cause: error });
                }
                // Anthropic Claude errors
                if (errorMessage.includes("does not support image") ||
                    errorMessage.includes("vision not supported")) {
                    throw new thread_input_error_1.ThreadInputError("This Claude model doesn't support images. Please use Claude 3.5 Sonnet, Claude 3 Opus, or other vision-capable models.", { cause: error });
                }
                // Generic image/vision errors
                if (errorMessage.includes("image") ||
                    errorMessage.includes("vision")) {
                    throw new thread_input_error_1.ThreadInputError("This model doesn't support image attachments. Please use a vision-capable model.", { cause: error });
                }
            }
            throw error;
        }
        // Clear text after successful submission
        setInputValue("");
    }, [
        inputValue,
        sendThreadMessage,
        thread.id,
        contextKey,
        imageState,
        mcpServers,
        resourceSource,
        clearInteractableSelections,
    ]);
    const { mutateAsync: submitAsync, mutate: _unusedSubmit, ...mutationState } = (0, react_query_hooks_1.useTamboMutation)({
        mutationFn: submit,
    });
    const value = {
        ...mutationState,
        value: inputValue,
        setValue: setInputValue,
        submit: submitAsync,
        images: imageState.images,
        addImage: imageState.addImage,
        addImages: imageState.addImages,
        removeImage: imageState.removeImage,
        clearImages: imageState.clearImages,
    };
    return (react_1.default.createElement(exports.TamboThreadInputContext.Provider, { value: value }, children));
};
exports.TamboThreadInputProvider = TamboThreadInputProvider;
/**
 * Hook to access the shared thread input state
 * contextKey parameter is not passed here anymore. Instead, use the contextKey prop in the TamboProvider.
 * @returns The thread input context
 */
const useTamboThreadInput = () => {
    const context = (0, react_1.useContext)(exports.TamboThreadInputContext);
    if (!context) {
        throw new Error("useTamboThreadInput must be used within a TamboThreadInputProvider");
    }
    return context;
};
exports.useTamboThreadInput = useTamboThreadInput;
