// @tambo-ai/react/dist/providers/tambo-stubs.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = require("@testing-library/react");
const react_2 = __importStar(require("react"));
const use_tambo_threads_1 = require("../hooks/use-tambo-threads");
const tambo_stubs_1 = require("./tambo-stubs");
describe("TamboStubProvider threads functionality", () => {
    const mockThread = {
        id: "test-thread",
        messages: [],
        createdAt: "2024-01-01T00:00:00Z",
        projectId: "test-project",
        updatedAt: "2024-01-01T00:00:00Z",
        metadata: {},
    };
    const mockThreadsData = {
        items: [mockThread],
        total: 1,
        count: 1,
        getPaginatedItems: () => [mockThread],
        nextPageRequestOptions: () => ({
            offset: 1,
            limit: 10,
            path: "/threads",
            method: "get",
        }),
    };
    it("should populate query cache with threads data", async () => {
        const wrapper = ({ children }) => (react_2.default.createElement(tambo_stubs_1.TamboStubProvider, { thread: mockThread, threads: mockThreadsData, projectId: "test-project", components: [], tools: [] }, children));
        const { result } = (0, react_1.renderHook)(() => (0, use_tambo_threads_1.useTamboThreadList)(), { wrapper });
        // Wait for the effect to run
        await (0, react_2.act)(async () => {
            await new Promise((resolve) => setTimeout(resolve, 0));
        });
        // The hook should return the pre-populated threads data
        expect(result.current.data).toEqual(mockThreadsData);
        expect(result.current.isLoading).toBe(false);
    });
    it("should populate query cache with threads data using contextKey", async () => {
        const wrapper = ({ children }) => (react_2.default.createElement(tambo_stubs_1.TamboStubProvider, { thread: mockThread, threads: mockThreadsData, projectId: "test-project", contextKey: "test-context", components: [], tools: [] }, children));
        const { result } = (0, react_1.renderHook)(() => (0, use_tambo_threads_1.useTamboThreadList)(), { wrapper });
        // Wait for the effect to run
        await (0, react_2.act)(async () => {
            await new Promise((resolve) => setTimeout(resolve, 0));
        });
        // The hook should return the pre-populated threads data for the specific context
        expect(result.current.data).toEqual(mockThreadsData);
        expect(result.current.isLoading).toBe(false);
    });
    it("should default projectId to thread.projectId when not provided", async () => {
        const wrapper = ({ children }) => (react_2.default.createElement(tambo_stubs_1.TamboStubProvider, { thread: mockThread, threads: mockThreadsData, 
            // No explicit projectId provided - should use thread.projectId
            components: [], tools: [] }, children));
        const { result } = (0, react_1.renderHook)(() => (0, use_tambo_threads_1.useTamboThreadList)(), { wrapper });
        // Wait for the effect to run
        await (0, react_2.act)(async () => {
            await new Promise((resolve) => setTimeout(resolve, 0));
        });
        // The hook should return the pre-populated threads data using thread.projectId
        expect(result.current.data).toEqual(mockThreadsData);
        expect(result.current.isLoading).toBe(false);
    });
});
