// @tambo-ai/react/dist/providers/tambo-prop-stream-provider/provider.js
"use strict";
"use client";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TamboPropStreamProviderComponent = exports.useTamboStream = void 0;
const react_1 = __importStar(require("react"));
const use_tambo_stream_status_1 = require("../../hooks/use-tambo-stream-status");
const types_1 = require("./types");
const TamboPropStreamContext = (0, react_1.createContext)(null);
/**
 * Hook to use the TamboStream context
 * @returns The TamboStream context
 */
const useTamboStream = () => {
    const context = (0, react_1.useContext)(TamboPropStreamContext);
    if (!context) {
        throw new Error("useTamboStream must be used within a TamboPropStreamProvider");
    }
    return context;
};
exports.useTamboStream = useTamboStream;
/**
 * The TamboStreamProvider provides a context for managing stream states
 * with compound components for Pending, Streaming, and Success states.
 * @param props - The props object
 * @param props.children - The children to wrap
 * @returns The TamboStreamProvider component
 */
const TamboPropStreamProviderComponent = ({ children, }) => {
    const { propStatus, streamStatus } = (0, use_tambo_stream_status_1.useTamboStreamStatus)();
    const keyStatusMap = (0, react_1.useMemo)(() => {
        const map = new Map();
        if (propStatus) {
            Object.entries(propStatus).forEach(([key, status]) => {
                map.set(key, {
                    isPending: status.isPending,
                    isStreaming: status.isStreaming,
                    isSuccess: status.isSuccess,
                    isError: !!status.error,
                    error: status.error,
                });
            });
        }
        // Always set default status for fallback
        map.set(types_1.DEFAULT_STREAM_KEY, {
            isPending: streamStatus.isPending,
            isStreaming: streamStatus.isStreaming,
            isSuccess: streamStatus.isSuccess,
            isError: streamStatus.isError,
            error: streamStatus.streamError,
        });
        return map;
    }, [streamStatus, propStatus]);
    const getStatusForKey = (0, react_1.useCallback)((key) => {
        // If the key exists in propStatus, return its status
        const propStatus = keyStatusMap.get(key);
        if (propStatus) {
            return propStatus;
        }
        // If key doesn't exist in propStatus, assume it's pending
        return {
            isPending: true,
            isStreaming: false,
            isSuccess: false,
            isError: false,
        };
    }, [keyStatusMap]);
    const contextValue = (0, react_1.useMemo)(() => ({
        streamStatus,
        getStatusForKey,
    }), [streamStatus, getStatusForKey]);
    return (react_1.default.createElement(TamboPropStreamContext.Provider, { value: contextValue }, children));
};
exports.TamboPropStreamProviderComponent = TamboPropStreamProviderComponent;
