// @tambo-ai/typescript-sdk/resources/beta/threads/messages.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../../core/resource.mjs";
import { buildHeaders } from "../../../internal/headers.mjs";
import { path } from "../../../internal/utils/path.mjs";
export class Messages extends APIResource {
    /**
     * @example
     * ```ts
     * const message = await client.beta.threads.messages.create(
     *   'id',
     *   { content: [{ type: 'text' }], role: 'user' },
     * );
     * ```
     */
    create(id, body, options) {
        return this._client.post(path `/threads/${id}/messages`, { body, ...options });
    }
    /**
     * @example
     * ```ts
     * const threadMessages =
     *   await client.beta.threads.messages.list('thr_123.456');
     * ```
     */
    list(id, query = {}, options) {
        return this._client.get(path `/threads/${id}/messages`, { query, ...options });
    }
    /**
     * @example
     * ```ts
     * await client.beta.threads.messages.delete('msg_123.456', {
     *   id: 'thr_123.456',
     * });
     * ```
     */
    delete(messageID, params, options) {
        const { id } = params;
        return this._client.delete(path `/threads/${id}/messages/${messageID}`, {
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * @example
     * ```ts
     * const threadMessage =
     *   await client.beta.threads.messages.updateComponentState(
     *     'msg_123.456',
     *     {
     *       id: 'thr_123.456',
     *       state: { foo: 'bar' },
     *     },
     *   );
     * ```
     */
    updateComponentState(messageID, params, options) {
        const { id, ...body } = params;
        return this._client.put(path `/threads/${id}/messages/${messageID}/component-state`, { body, ...options });
    }
}
