// @tambo-ai/typescript-sdk/resources/beta/threads/messages.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Messages = void 0;
const resource_1 = require("../../../core/resource.js");
const headers_1 = require("../../../internal/headers.js");
const path_1 = require("../../../internal/utils/path.js");
class Messages extends resource_1.APIResource {
    /**
     * @example
     * ```ts
     * const message = await client.beta.threads.messages.create(
     *   'id',
     *   { content: [{ type: 'text' }], role: 'user' },
     * );
     * ```
     */
    create(id, body, options) {
        return this._client.post((0, path_1.path) `/threads/${id}/messages`, { body, ...options });
    }
    /**
     * @example
     * ```ts
     * const threadMessages =
     *   await client.beta.threads.messages.list('thr_123.456');
     * ```
     */
    list(id, query = {}, options) {
        return this._client.get((0, path_1.path) `/threads/${id}/messages`, { query, ...options });
    }
    /**
     * @example
     * ```ts
     * await client.beta.threads.messages.delete('msg_123.456', {
     *   id: 'thr_123.456',
     * });
     * ```
     */
    delete(messageID, params, options) {
        const { id } = params;
        return this._client.delete((0, path_1.path) `/threads/${id}/messages/${messageID}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * @example
     * ```ts
     * const threadMessage =
     *   await client.beta.threads.messages.updateComponentState(
     *     'msg_123.456',
     *     {
     *       id: 'thr_123.456',
     *       state: { foo: 'bar' },
     *     },
     *   );
     * ```
     */
    updateComponentState(messageID, params, options) {
        const { id, ...body } = params;
        return this._client.put((0, path_1.path) `/threads/${id}/messages/${messageID}/component-state`, { body, ...options });
    }
}
exports.Messages = Messages;
