// @tambo-ai/typescript-sdk/resources/beta/threads/threads.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../../core/resource.mjs";
import * as MessagesAPI from "./messages.mjs";
import { Messages, } from "./messages.mjs";
import * as SuggestionsAPI from "./suggestions.mjs";
import { Suggestions, } from "./suggestions.mjs";
import { OffsetAndLimit } from "../../../core/pagination.mjs";
import { buildHeaders } from "../../../internal/headers.mjs";
import { path } from "../../../internal/utils/path.mjs";
export class Threads extends APIResource {
    constructor() {
        super(...arguments);
        this.messages = new MessagesAPI.Messages(this._client);
        this.suggestions = new SuggestionsAPI.Suggestions(this._client);
    }
    /**
     * @example
     * ```ts
     * const thread = await client.beta.threads.create({
     *   projectId: 'projectId',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/threads', { body, ...options });
    }
    /**
     * @example
     * ```ts
     * const thread = await client.beta.threads.retrieve('id');
     * ```
     */
    retrieve(id, query = {}, options) {
        return this._client.get(path `/threads/${id}`, { query, ...options });
    }
    /**
     * @example
     * ```ts
     * const thread = await client.beta.threads.update('id', {
     *   projectId: 'projectId',
     * });
     * ```
     */
    update(id, body, options) {
        return this._client.put(path `/threads/${id}`, { body, ...options });
    }
    /**
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const thread of client.beta.threads.list(
     *   'projectId',
     * )) {
     *   // ...
     * }
     * ```
     */
    list(projectID, query = {}, options) {
        return this._client.getAPIList(path `/threads/project/${projectID}`, (OffsetAndLimit), {
            query,
            ...options,
        });
    }
    /**
     * @example
     * ```ts
     * await client.beta.threads.delete('id');
     * ```
     */
    delete(id, options) {
        return this._client.delete(path `/threads/${id}`, {
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * This endpoint has been deprecated. Use POST /advancestream instead.
     *
     * @deprecated
     */
    advance(options) {
        return this._client.post('/threads/advance', options);
    }
    /**
     * This endpoint has been deprecated. Use POST /:id/advancestream instead.
     *
     * @deprecated
     */
    advanceByID(id, body, options) {
        return this._client.post(path `/threads/${id}/advance`, { body, ...options });
    }
    /**
     * Sets a thread's generation stage to CANCELLED
     *
     * @example
     * ```ts
     * const thread = await client.beta.threads.cancel(
     *   'thread_123456789',
     * );
     * ```
     */
    cancel(id, options) {
        return this._client.post(path `/threads/${id}/cancel`, options);
    }
    /**
     * Automatically generates and sets a name for the thread as a summary based on its
     * messages.
     *
     * @example
     * ```ts
     * const thread = await client.beta.threads.generateName(
     *   'thread_123456789',
     * );
     * ```
     */
    generateName(id, params = {}, options) {
        const { contextKey } = params ?? {};
        return this._client.post(path `/threads/${id}/generate-name`, { query: { contextKey }, ...options });
    }
}
Threads.Messages = Messages;
Threads.Suggestions = Suggestions;
