// @tambo-ai/typescript-sdk/resources/beta/threads/threads.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Threads = void 0;
const tslib_1 = require("../../../internal/tslib.js");
const resource_1 = require("../../../core/resource.js");
const MessagesAPI = tslib_1.__importStar(require("./messages.js"));
const messages_1 = require("./messages.js");
const SuggestionsAPI = tslib_1.__importStar(require("./suggestions.js"));
const suggestions_1 = require("./suggestions.js");
const pagination_1 = require("../../../core/pagination.js");
const headers_1 = require("../../../internal/headers.js");
const path_1 = require("../../../internal/utils/path.js");
class Threads extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.messages = new MessagesAPI.Messages(this._client);
        this.suggestions = new SuggestionsAPI.Suggestions(this._client);
    }
    /**
     * @example
     * ```ts
     * const thread = await client.beta.threads.create({
     *   projectId: 'projectId',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/threads', { body, ...options });
    }
    /**
     * @example
     * ```ts
     * const thread = await client.beta.threads.retrieve('id');
     * ```
     */
    retrieve(id, query = {}, options) {
        return this._client.get((0, path_1.path) `/threads/${id}`, { query, ...options });
    }
    /**
     * @example
     * ```ts
     * const thread = await client.beta.threads.update('id', {
     *   projectId: 'projectId',
     * });
     * ```
     */
    update(id, body, options) {
        return this._client.put((0, path_1.path) `/threads/${id}`, { body, ...options });
    }
    /**
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const thread of client.beta.threads.list(
     *   'projectId',
     * )) {
     *   // ...
     * }
     * ```
     */
    list(projectID, query = {}, options) {
        return this._client.getAPIList((0, path_1.path) `/threads/project/${projectID}`, (pagination_1.OffsetAndLimit), {
            query,
            ...options,
        });
    }
    /**
     * @example
     * ```ts
     * await client.beta.threads.delete('id');
     * ```
     */
    delete(id, options) {
        return this._client.delete((0, path_1.path) `/threads/${id}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * This endpoint has been deprecated. Use POST /advancestream instead.
     *
     * @deprecated
     */
    advance(options) {
        return this._client.post('/threads/advance', options);
    }
    /**
     * This endpoint has been deprecated. Use POST /:id/advancestream instead.
     *
     * @deprecated
     */
    advanceByID(id, body, options) {
        return this._client.post((0, path_1.path) `/threads/${id}/advance`, { body, ...options });
    }
    /**
     * Sets a thread's generation stage to CANCELLED
     *
     * @example
     * ```ts
     * const thread = await client.beta.threads.cancel(
     *   'thread_123456789',
     * );
     * ```
     */
    cancel(id, options) {
        return this._client.post((0, path_1.path) `/threads/${id}/cancel`, options);
    }
    /**
     * Automatically generates and sets a name for the thread as a summary based on its
     * messages.
     *
     * @example
     * ```ts
     * const thread = await client.beta.threads.generateName(
     *   'thread_123456789',
     * );
     * ```
     */
    generateName(id, params = {}, options) {
        const { contextKey } = params ?? {};
        return this._client.post((0, path_1.path) `/threads/${id}/generate-name`, { query: { contextKey }, ...options });
    }
}
exports.Threads = Threads;
Threads.Messages = messages_1.Messages;
Threads.Suggestions = suggestions_1.Suggestions;
