// @tambo-ai/typescript-sdk/src/resources/shared.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

export interface AvailableComponent {
  contextTools: Array<ComponentContextToolMetadata>;

  description: string;

  name: string;

  /**
   * JSON Schema for the component's props
   */
  props: { [key: string]: unknown };
}

export interface ComponentContextToolMetadata {
  description: string;

  name: string;

  parameters: Array<ToolParameters>;

  /**
   * Maximum number of times this tool can be called in a single thread
   * execution.Overrides project-level maxToolCalls setting for this specific tool.
   */
  maxCalls?: number;
}

export interface Resource {
  /**
   * Annotations for additional metadata (MCP-specific). Can include audience,
   * priority, or custom properties.
   */
  annotations?: ResourceAnnotations;

  /**
   * Base64-encoded blob data (alternative to uri or text)
   */
  blob?: string;

  /**
   * Optional description of the resource
   */
  description?: string;

  /**
   * MIME type of the resource
   */
  mimeType?: string;

  /**
   * Human-readable name for the resource
   */
  name?: string;

  /**
   * Inline text content (alternative to uri)
   */
  text?: string;

  /**
   * URI identifying the resource (e.g., file://, https://, s3://)
   */
  uri?: string;
}

export interface ResourceAnnotations {
  /**
   * Target audience for this resource
   */
  audience?: Array<string>;

  /**
   * Priority level for this resource
   */
  priority?: number;
}

export interface ToolCallParameter {
  parameterName: string;

  parameterValue: unknown;
}

export interface ToolCallRequest {
  /**
   * The parameters of the function to call.
   */
  parameters: Array<ToolCallParameter>;

  /**
   * The name of the function to call.
   */
  toolName: string;

  /**
   * @deprecated The unique id of the tool call, no longer used.
   */
  tool_call_id?: string;
}

export interface ToolParameters {
  description: string;

  isRequired: boolean;

  name: string;

  type: string;

  enumValues?: Array<string>;

  items?: ToolParameters.Items;

  schema?: unknown;
}

export namespace ToolParameters {
  export interface Items {
    type: string;
  }
}
