// @tambo-ai/typescript-sdk/src/resources/storage.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import { APIResource } from '../core/resource';
import { APIPromise } from '../core/api-promise';
import { RequestOptions } from '../internal/request-options';

export class Storage extends APIResource {
  /**
   * Returns a presigned URL that allows direct file upload to S3. The client should
   * PUT the file directly to the returned uploadUrl.
   *
   * @example
   * ```ts
   * const response = await client.storage.presign({
   *   contentType: 'application/pdf',
   *   size: 1024000,
   * });
   * ```
   */
  presign(body: StoragePresignParams, options?: RequestOptions): APIPromise<StoragePresignResponse> {
    return this._client.post('/storage/presign', { body, ...options });
  }
}

export interface StoragePresignResponse {
  /**
   * The attachment URI to reference this file in messages
   */
  attachmentUri: string;

  /**
   * Time in seconds until the presigned URL expires
   */
  expiresIn: number;

  /**
   * The presigned URL to PUT the file directly to S3
   */
  uploadUrl: string;
}

export interface StoragePresignParams {
  /**
   * The MIME type of the file
   */
  contentType: string;

  /**
   * The file size in bytes (max 10MB)
   */
  size: number;
}

export declare namespace Storage {
  export {
    type StoragePresignResponse as StoragePresignResponse,
    type StoragePresignParams as StoragePresignParams,
  };
}
