// @tambo-ai/typescript-sdk/src/resources/beta/threads/messages.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import { APIResource } from '../../../core/resource';
import * as Shared from '../../shared';
import * as ThreadsAPI from './threads';
import { APIPromise } from '../../../core/api-promise';
import { buildHeaders } from '../../../internal/headers';
import { RequestOptions } from '../../../internal/request-options';
import { path } from '../../../internal/utils/path';

export class Messages extends APIResource {
  /**
   * @example
   * ```ts
   * const message = await client.beta.threads.messages.create(
   *   'id',
   *   { content: [{ type: 'text' }], role: 'user' },
   * );
   * ```
   */
  create(id: string, body: MessageCreateParams, options?: RequestOptions): APIPromise<unknown> {
    return this._client.post(path`/threads/${id}/messages`, { body, ...options });
  }

  /**
   * @example
   * ```ts
   * const threadMessages =
   *   await client.beta.threads.messages.list('thr_123.456');
   * ```
   */
  list(
    id: string,
    query: MessageListParams | null | undefined = {},
    options?: RequestOptions,
  ): APIPromise<MessageListResponse> {
    return this._client.get(path`/threads/${id}/messages`, { query, ...options });
  }

  /**
   * @example
   * ```ts
   * await client.beta.threads.messages.delete('msg_123.456', {
   *   id: 'thr_123.456',
   * });
   * ```
   */
  delete(messageID: string, params: MessageDeleteParams, options?: RequestOptions): APIPromise<void> {
    const { id } = params;
    return this._client.delete(path`/threads/${id}/messages/${messageID}`, {
      ...options,
      headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
    });
  }

  /**
   * @example
   * ```ts
   * const threadMessage =
   *   await client.beta.threads.messages.updateComponentState(
   *     'msg_123.456',
   *     {
   *       id: 'thr_123.456',
   *       state: { foo: 'bar' },
   *     },
   *   );
   * ```
   */
  updateComponentState(
    messageID: string,
    params: MessageUpdateComponentStateParams,
    options?: RequestOptions,
  ): APIPromise<ThreadsAPI.ThreadMessage> {
    const { id, ...body } = params;
    return this._client.put(path`/threads/${id}/messages/${messageID}/component-state`, { body, ...options });
  }
}

export type MessageCreateResponse = unknown;

export type MessageListResponse = Array<ThreadsAPI.ThreadMessage>;

export interface MessageCreateParams {
  content: Array<ThreadsAPI.ChatCompletionContentPart>;

  role: 'user' | 'assistant' | 'system' | 'tool';

  /**
   * @deprecated No longer used - instead set role and the tool call information
   */
  actionType?: 'tool_call' | 'tool_response';

  /**
   * Additional context to provide to the AI beyond the user query, such as the info
   * about the current page the user is visiting.
   */
  additionalContext?: { [key: string]: unknown };

  component?: ThreadsAPI.ComponentDecisionV2;

  /**
   * The initial state of the component
   */
  componentState?: { [key: string]: unknown };

  error?: string;

  /**
   * Whether the message has been cancelled
   */
  isCancelled?: boolean;

  metadata?: { [key: string]: unknown };

  /**
   * Reasoning text from the LLM, if the model supports it.
   */
  reasoning?: Array<string>;

  /**
   * Duration of reasoning in milliseconds
   */
  reasoningDurationMS?: number;

  tool_call_id?: string;

  toolCallRequest?: Shared.ToolCallRequest;

  /**
   * @deprecated Put the response in the content instead
   */
  toolResponse?: unknown;
}

export interface MessageListParams {
  /**
   * Whether to include internal messages, must be `true`
   */
  includeInternal?: boolean;
}

export interface MessageDeleteParams {
  /**
   * Id of the thread that contains the message
   */
  id: string;
}

export interface MessageUpdateComponentStateParams {
  /**
   * Path param: Id of the thread that contains the message
   */
  id: string;

  /**
   * Body param
   */
  state: { [key: string]: unknown };
}

export declare namespace Messages {
  export {
    type MessageCreateResponse as MessageCreateResponse,
    type MessageListResponse as MessageListResponse,
    type MessageCreateParams as MessageCreateParams,
    type MessageListParams as MessageListParams,
    type MessageDeleteParams as MessageDeleteParams,
    type MessageUpdateComponentStateParams as MessageUpdateComponentStateParams,
  };
}
