// @tambo-ai/typescript-sdk/src/resources/beta/threads/suggestions.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import { APIResource } from '../../../core/resource';
import * as Shared from '../../shared';
import { APIPromise } from '../../../core/api-promise';
import { RequestOptions } from '../../../internal/request-options';
import { path } from '../../../internal/utils/path';

export class Suggestions extends APIResource {
  /**
   * Retrieves all suggestions generated for a specific message
   *
   * @example
   * ```ts
   * const suggestions =
   *   await client.beta.threads.suggestions.list(
   *     'msg_123456789',
   *     { id: 'thread_123456789' },
   *   );
   * ```
   */
  list(
    messageID: string,
    params: SuggestionListParams,
    options?: RequestOptions,
  ): APIPromise<SuggestionListResponse> {
    const { id } = params;
    return this._client.get(path`/threads/${id}/messages/${messageID}/suggestions`, options);
  }

  /**
   * Generates and stores new suggestions for a specific message
   *
   * @example
   * ```ts
   * const suggestions =
   *   await client.beta.threads.suggestions.generate(
   *     'msg_123456789',
   *     { id: 'thread_123456789' },
   *   );
   * ```
   */
  generate(
    messageID: string,
    params: SuggestionGenerateParams,
    options?: RequestOptions,
  ): APIPromise<SuggestionGenerateResponse> {
    const { id, ...body } = params;
    return this._client.post(path`/threads/${id}/messages/${messageID}/suggestions`, { body, ...options });
  }
}

export interface Suggestion {
  /**
   * Unique identifier for the suggestion
   */
  id: string;

  /**
   * Detailed explanation of the suggestion
   */
  detailedSuggestion: string;

  /**
   * ID of the message this suggestion is for
   */
  messageId: string;

  /**
   * Short title or summary of the suggestion
   */
  title: string;

  /**
   * Additional metadata for the suggestion
   */
  metadata?: { [key: string]: string };
}

export interface SuggestionsGenerate {
  /**
   * Available components that can be used with this suggestion
   */
  availableComponents?: Array<Shared.AvailableComponent>;

  /**
   * Maximum number of suggestions to generate
   */
  maxSuggestions?: number;
}

export type SuggestionListResponse = Array<Suggestion>;

export type SuggestionGenerateResponse = Array<Suggestion>;

export interface SuggestionListParams {
  /**
   * ID of the thread to get suggestions for
   */
  id: string;
}

export interface SuggestionGenerateParams {
  /**
   * Path param: ID of the thread to generate suggestions for
   */
  id: string;

  /**
   * Body param: Available components that can be used with this suggestion
   */
  availableComponents?: Array<Shared.AvailableComponent>;

  /**
   * Body param: Maximum number of suggestions to generate
   */
  maxSuggestions?: number;
}

export declare namespace Suggestions {
  export {
    type Suggestion as Suggestion,
    type SuggestionsGenerate as SuggestionsGenerate,
    type SuggestionListResponse as SuggestionListResponse,
    type SuggestionGenerateResponse as SuggestionGenerateResponse,
    type SuggestionListParams as SuggestionListParams,
    type SuggestionGenerateParams as SuggestionGenerateParams,
  };
}
