// @tambo-ai/react/dist/hooks/use-current-message.js
"use strict";
"use client";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTamboCurrentComponent = exports.useTamboCurrentMessage = exports.TamboMessageProvider = exports.TamboMessageContext = void 0;
exports.wrapWithTamboMessageProvider = wrapWithTamboMessageProvider;
const react_1 = __importStar(require("react"));
exports.TamboMessageContext = (0, react_1.createContext)(null);
/**
 * Wraps all components, so that they can find what message they are in.
 * Also supports optional interactable metadata for components wrapped with withInteractable.
 * @param props - props for the TamboMessageProvider
 * @param props.children - The children to wrap
 * @param props.message - The message object
 * @param props.interactableMetadata - Optional interactable component metadata
 * @returns The wrapped component
 */
const TamboMessageProvider = ({ children, message, interactableMetadata, }) => {
    // Merge interactable metadata into message if provided
    const enhancedMessage = interactableMetadata
        ? { ...message, interactableMetadata }
        : message;
    // Use a unique key={...} to force a re-render when the message changes - this
    // make sure that if the rendered component is swapped into a tree (like if
    // you always show the last rendered component) then the state/etc is correct
    return (react_1.default.createElement(exports.TamboMessageContext.Provider, { value: enhancedMessage, key: message.id }, children));
};
exports.TamboMessageProvider = TamboMessageProvider;
/**
 * Wraps a component with a TamboMessageProvider - this allows the provider
 * to be used outside of a TSX file
 * @param children - The children to wrap
 * @param message - The message object
 * @param interactableMetadata - Optional interactable metadata
 * @returns The wrapped component
 */
function wrapWithTamboMessageProvider(children, message, interactableMetadata) {
    return (react_1.default.createElement(exports.TamboMessageProvider, { message: message, interactableMetadata: interactableMetadata }, children));
}
/**
 * Hook used inside a component wrapped with TamboMessageProvider, to get
 * the current message.
 * @returns The current message that is used to render the component
 */
const useTamboCurrentMessage = () => {
    const message = (0, react_1.useContext)(exports.TamboMessageContext);
    if (!message) {
        throw new Error("useTamboCurrentMessage must be used within a TamboMessageProvider");
    }
    return message;
};
exports.useTamboCurrentMessage = useTamboCurrentMessage;
/**
 * Hook to access the current component information from the message context.
 * Provides a unified interface for both AI-generated and interactable components.
 *
 * **Use this hook when you need component metadata regardless of the context.**
 * @returns Component info including name, props, and interactable metadata if available, or null if used outside TamboMessageProvider
 * @example
 * ```tsx
 * function MyInlineEditor() {
 *   const component = useTamboCurrentComponent();
 *
 *   if (!component) return null; // Not inside a component
 *
 *   return (
 *     <div>
 *       Editing: {component.componentName}
 *       {component.interactableId && <span>ID: {component.interactableId}</span>}
 *     </div>
 *   );
 * }
 * ```
 */
const useTamboCurrentComponent = () => {
    const message = (0, react_1.useContext)(exports.TamboMessageContext);
    if (!message) {
        return null;
    }
    return {
        componentName: message.interactableMetadata?.componentName ??
            message.component?.componentName ??
            undefined,
        props: message.component?.props,
        interactableId: message.interactableMetadata?.id ?? undefined,
        description: message.interactableMetadata?.description ?? undefined,
        threadId: message.threadId,
    };
};
exports.useTamboCurrentComponent = useTamboCurrentComponent;
