// @tambo-ai/react/dist/hooks/use-tambo-voice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTamboVoice = useTamboVoice;
const react_1 = require("react");
const react_media_recorder_1 = require("react-media-recorder");
const tambo_client_provider_1 = require("../providers/tambo-client-provider");
const react_query_hooks_1 = require("./react-query-hooks");
/**
 * Exposes functionality to record speech and transcribe it using the Tambo API.
 * @returns An object with:
 * - startRecording: A function to start recording audio and reset the current transcript.
 * - stopRecording: A function to stop recording audio and automatically kick off transcription.
 * - isRecording: A boolean indicating if the user is recording audio.
 * - isTranscribing: A boolean indicating if the audio is being transcribed.
 * - transcript: The transcript of the recorded audio.
 * - transcriptionError: An error message if the transcription fails.
 * - mediaAccessError: An error message if microphone access fails.
 */
function useTamboVoice() {
    const [transcript, setTranscript] = (0, react_1.useState)(null);
    const client = (0, tambo_client_provider_1.useTamboClient)();
    const { status, startRecording: startMediaRecording, stopRecording: stopMediaRecording, mediaBlobUrl, error: mediaAccessError, } = (0, react_media_recorder_1.useReactMediaRecorder)({
        audio: true,
        video: false,
        blobPropertyBag: { type: "audio/webm" },
    });
    const isRecording = status === "recording";
    const transcriptionMutation = (0, react_query_hooks_1.useTamboMutation)({
        mutationFn: async (blobUrl) => {
            const response = await fetch(blobUrl);
            const audioBlob = await response.blob();
            const file = new File([audioBlob], "recording.webm", {
                type: "audio/webm",
            });
            return await client.beta.audio.transcribe({ file });
        },
        onSuccess: (transcription) => {
            setTranscript(transcription);
        },
    });
    // Trigger transcription when recording stops and we have a blob URL
    const shouldTranscribe = status === "stopped" &&
        mediaBlobUrl &&
        !transcriptionMutation.isPending &&
        !transcriptionMutation.isSuccess;
    (0, react_1.useEffect)(() => {
        if (shouldTranscribe) {
            transcriptionMutation.mutate(mediaBlobUrl);
        }
    }, [shouldTranscribe, mediaBlobUrl, transcriptionMutation]);
    const startRecording = (0, react_1.useCallback)(() => {
        if (isRecording)
            return;
        // Reset state when starting new recording
        setTranscript(null);
        transcriptionMutation.reset(); // Clear any previous mutation state
        startMediaRecording();
    }, [isRecording, startMediaRecording, transcriptionMutation]);
    const stopRecording = (0, react_1.useCallback)(() => {
        if (isRecording) {
            stopMediaRecording();
        }
    }, [isRecording, stopMediaRecording]);
    return {
        startRecording,
        stopRecording,
        isRecording,
        mediaAccessError: mediaAccessError === "" ? null : mediaAccessError,
        isTranscribing: transcriptionMutation.isPending,
        transcript,
        transcriptionError: transcriptionMutation.error?.message ?? null,
    };
}
