// @tambo-ai/react/esm/providers/tambo-component-provider.js
"use client";
import React, { createContext, useContext } from "react";
import { useTamboClient } from "./tambo-client-provider";
import { useTamboRegistry } from "./tambo-registry-provider";
const TamboComponentContext = createContext({
    /**
     *
     */
    registerComponent: () => { },
    /**
     *
     */
    registerTool: () => { },
    /**
     *
     */
    registerTools: () => { },
    /**
     *
     */
    addToolAssociation: () => { },
});
/**
 * The TamboComponentProvider is a React provider that provides component
 * registration services to the descendants of the provider.
 * @param props - The props for the TamboComponentProvider
 * @param props.children - The children to wrap
 * @returns The TamboComponentProvider component
 */
export const TamboComponentProvider = ({ children, }) => {
    const client = useTamboClient();
    const { registerComponent, addToolAssociation, registerTool, registerTools } = useTamboRegistry();
    const value = {
        client,
        registerComponent,
        registerTool,
        registerTools,
        addToolAssociation,
    };
    return (React.createElement(TamboComponentContext.Provider, { value: value }, children));
};
/**
 * The useTamboComponent hook provides access to the component registration
 * services to the descendants of the TamboComponentProvider.
 * @returns The component registration services
 */
export const useTamboComponent = () => {
    return useContext(TamboComponentContext);
};
