// @tambo-ai/typescript-sdk/src/resources/beta/beta.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import { APIResource } from '../../core/resource';
import * as AudioAPI from './audio';
import { Audio, AudioTranscribeParams, AudioTranscribeResponse } from './audio';
import * as AuthAPI from './auth';
import {
  Auth,
  AuthGetMcpTokenParams,
  AuthGetMcpTokenResponse,
  AuthGetTokenParams,
  AuthGetTokenResponse,
} from './auth';
import * as RegistryAPI from './registry';
import { Registry, RegistryRetrieveResponse } from './registry';
import * as ProjectsAPI from './projects/projects';
import {
  ProjectDeleteResponse,
  ProjectGetCurrentResponse,
  ProjectRetrieveResponse,
  Projects,
} from './projects/projects';
import * as ThreadsAPI from './threads/threads';
import {
  ChatCompletionContentPart,
  ComponentDecisionV2,
  Thread,
  ThreadAdvanceByIDParams,
  ThreadAdvanceByIDResponse,
  ThreadAdvanceResponse,
  ThreadCreateParams,
  ThreadGenerateNameParams,
  ThreadListParams,
  ThreadMessage,
  ThreadRetrieveParams,
  ThreadRetrieveResponse,
  ThreadUpdateParams,
  Threads,
  ThreadsOffsetAndLimit,
} from './threads/threads';

export class Beta extends APIResource {
  registry: RegistryAPI.Registry = new RegistryAPI.Registry(this._client);
  projects: ProjectsAPI.Projects = new ProjectsAPI.Projects(this._client);
  threads: ThreadsAPI.Threads = new ThreadsAPI.Threads(this._client);
  auth: AuthAPI.Auth = new AuthAPI.Auth(this._client);
  audio: AudioAPI.Audio = new AudioAPI.Audio(this._client);
}

Beta.Registry = Registry;
Beta.Projects = Projects;
Beta.Threads = Threads;
Beta.Auth = Auth;
Beta.Audio = Audio;

export declare namespace Beta {
  export { Registry as Registry, type RegistryRetrieveResponse as RegistryRetrieveResponse };

  export {
    Projects as Projects,
    type ProjectRetrieveResponse as ProjectRetrieveResponse,
    type ProjectDeleteResponse as ProjectDeleteResponse,
    type ProjectGetCurrentResponse as ProjectGetCurrentResponse,
  };

  export {
    Threads as Threads,
    type ChatCompletionContentPart as ChatCompletionContentPart,
    type ComponentDecisionV2 as ComponentDecisionV2,
    type Thread as Thread,
    type ThreadMessage as ThreadMessage,
    type ThreadRetrieveResponse as ThreadRetrieveResponse,
    type ThreadAdvanceResponse as ThreadAdvanceResponse,
    type ThreadAdvanceByIDResponse as ThreadAdvanceByIDResponse,
    type ThreadsOffsetAndLimit as ThreadsOffsetAndLimit,
    type ThreadCreateParams as ThreadCreateParams,
    type ThreadRetrieveParams as ThreadRetrieveParams,
    type ThreadUpdateParams as ThreadUpdateParams,
    type ThreadListParams as ThreadListParams,
    type ThreadAdvanceByIDParams as ThreadAdvanceByIDParams,
    type ThreadGenerateNameParams as ThreadGenerateNameParams,
  };

  export {
    Auth as Auth,
    type AuthGetMcpTokenResponse as AuthGetMcpTokenResponse,
    type AuthGetTokenResponse as AuthGetTokenResponse,
    type AuthGetMcpTokenParams as AuthGetMcpTokenParams,
    type AuthGetTokenParams as AuthGetTokenParams,
  };

  export {
    Audio as Audio,
    type AudioTranscribeResponse as AudioTranscribeResponse,
    type AudioTranscribeParams as AudioTranscribeParams,
  };
}
