// @tambo-ai/react/dist/schema/json-schema.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const json_schema_1 = require("./json-schema");
describe("looksLikeJSONSchema", () => {
    it("should return true for a valid object schema", () => {
        const schema = {
            type: "object",
            properties: {
                name: { type: "string" },
            },
        };
        expect((0, json_schema_1.looksLikeJSONSchema)(schema)).toBe(true);
    });
    it("should return true for a schema with only properties", () => {
        const schema = {
            properties: {
                name: { type: "string" },
            },
        };
        expect((0, json_schema_1.looksLikeJSONSchema)(schema)).toBe(true);
    });
    it("should return true for a schema with items (array)", () => {
        const schema = {
            type: "array",
            items: { type: "string" },
        };
        expect((0, json_schema_1.looksLikeJSONSchema)(schema)).toBe(true);
    });
    it("should return true for a schema with enum", () => {
        const schema = {
            enum: ["a", "b", "c"],
        };
        expect((0, json_schema_1.looksLikeJSONSchema)(schema)).toBe(true);
    });
    it("should return true for a schema with const", () => {
        const schema = {
            const: "fixed-value",
        };
        expect((0, json_schema_1.looksLikeJSONSchema)(schema)).toBe(true);
    });
    it("should return true for a schema with $ref", () => {
        const schema = {
            $ref: "#/definitions/Person",
        };
        expect((0, json_schema_1.looksLikeJSONSchema)(schema)).toBe(true);
    });
    it("should return false for null", () => {
        expect((0, json_schema_1.looksLikeJSONSchema)(null)).toBe(false);
    });
    it("should return false for non-object", () => {
        expect((0, json_schema_1.looksLikeJSONSchema)("string")).toBe(false);
        expect((0, json_schema_1.looksLikeJSONSchema)(123)).toBe(false);
        expect((0, json_schema_1.looksLikeJSONSchema)(undefined)).toBe(false);
    });
    it("should return false for empty object", () => {
        expect((0, json_schema_1.looksLikeJSONSchema)({})).toBe(false);
    });
    it("should return false for object without schema keys", () => {
        const notSchema = {
            foo: "bar",
            baz: 123,
        };
        expect((0, json_schema_1.looksLikeJSONSchema)(notSchema)).toBe(false);
    });
});
describe("isJsonSchemaTuple", () => {
    it("should return true for draft-07 tuple (items as array)", () => {
        const schema = {
            type: "array",
            items: [{ type: "string" }, { type: "number" }],
        };
        expect((0, json_schema_1.isJsonSchemaTuple)(schema)).toBe(true);
    });
    it("should return true for draft 2020-12 tuple (prefixItems)", () => {
        const schema = {
            type: "array",
            prefixItems: [{ type: "string" }, { type: "number" }],
        };
        expect((0, json_schema_1.isJsonSchemaTuple)(schema)).toBe(true);
    });
    it("should return false for regular array schema (items as object)", () => {
        const schema = {
            type: "array",
            items: { type: "string" },
        };
        expect((0, json_schema_1.isJsonSchemaTuple)(schema)).toBe(false);
    });
    it("should return false for non-array schema", () => {
        const schema = {
            type: "object",
            properties: {},
        };
        expect((0, json_schema_1.isJsonSchemaTuple)(schema)).toBe(false);
    });
    it("should return false for array without items or prefixItems", () => {
        const schema = {
            type: "array",
        };
        expect((0, json_schema_1.isJsonSchemaTuple)(schema)).toBe(false);
    });
});
describe("getJsonSchemaTupleItems", () => {
    it("should return items for draft-07 tuple", () => {
        const items = [{ type: "string" }, { type: "number" }];
        const schema = {
            type: "array",
            items: items,
        };
        expect((0, json_schema_1.getJsonSchemaTupleItems)(schema)).toEqual(items);
    });
    it("should return prefixItems for draft 2020-12 tuple", () => {
        const prefixItems = [{ type: "string" }, { type: "number" }];
        const schema = {
            type: "array",
            prefixItems: prefixItems,
        };
        expect((0, json_schema_1.getJsonSchemaTupleItems)(schema)).toEqual(prefixItems);
    });
    it("should return undefined for non-array schema", () => {
        const schema = {
            type: "object",
        };
        expect((0, json_schema_1.getJsonSchemaTupleItems)(schema)).toBeUndefined();
    });
    it("should return undefined for array without tuple items", () => {
        const schema = {
            type: "array",
            items: { type: "string" },
        };
        expect((0, json_schema_1.getJsonSchemaTupleItems)(schema)).toBeUndefined();
    });
    it("should prefer prefixItems over items when both present", () => {
        const prefixItems = [{ type: "boolean" }];
        const items = [{ type: "string" }];
        const schema = {
            type: "array",
            prefixItems: prefixItems,
            items: items,
        };
        expect((0, json_schema_1.getJsonSchemaTupleItems)(schema)).toEqual(prefixItems);
    });
});
describe("makeJsonSchemaPartial", () => {
    it("should remove required array from schema", () => {
        const schema = {
            type: "object",
            properties: {
                name: { type: "string" },
                age: { type: "number" },
            },
            required: ["name", "age"],
        };
        const partial = (0, json_schema_1.makeJsonSchemaPartial)(schema);
        expect(partial.required).toBeUndefined();
        expect(partial.type).toBe("object");
        expect(partial.properties).toEqual(schema.properties);
    });
    it("should preserve all other properties", () => {
        const schema = {
            type: "object",
            properties: {
                name: { type: "string", description: "The name" },
            },
            required: ["name"],
            additionalProperties: false,
            description: "A person schema",
            title: "Person",
        };
        const partial = (0, json_schema_1.makeJsonSchemaPartial)(schema);
        expect(partial.required).toBeUndefined();
        expect(partial.type).toBe("object");
        expect(partial.properties).toEqual(schema.properties);
        expect(partial.additionalProperties).toBe(false);
        expect(partial.description).toBe("A person schema");
        expect(partial.title).toBe("Person");
    });
    it("should handle schema without required array", () => {
        const schema = {
            type: "object",
            properties: {
                name: { type: "string" },
            },
        };
        const partial = (0, json_schema_1.makeJsonSchemaPartial)(schema);
        expect(partial.required).toBeUndefined();
        expect(partial.type).toBe("object");
        expect(partial.properties).toEqual(schema.properties);
    });
    it("should not mutate the original schema", () => {
        const schema = {
            type: "object",
            properties: {
                name: { type: "string" },
            },
            required: ["name"],
        };
        const partial = (0, json_schema_1.makeJsonSchemaPartial)(schema);
        // Original should be unchanged
        expect(schema.required).toEqual(["name"]);
        // Partial should not have required
        expect(partial.required).toBeUndefined();
    });
});
