// @tambo-ai/react/dist/util/resource-validators.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const resource_validators_1 = require("./resource-validators");
describe("validateResourceSource", () => {
    it("should throw when has listResources but not getResource", () => {
        const listResources = async () => [];
        const getResource = undefined;
        expect(() => {
            (0, resource_validators_1.validateResourceSource)(listResources, getResource);
        }).toThrow("Both listResources and getResource must be provided together, or neither should be provided. " +
            "Got: listResources=defined, getResource=undefined");
    });
    it("should throw when has getResource but not listResources", () => {
        const listResources = undefined;
        const getResource = async () => ({
            contents: [
                {
                    uri: "https://resource",
                    mimeType: "text/plain",
                    text: "test",
                },
            ],
        });
        expect(() => {
            (0, resource_validators_1.validateResourceSource)(listResources, getResource);
        }).toThrow("Both listResources and getResource must be provided together, or neither should be provided. " +
            "Got: listResources=undefined, getResource=defined");
    });
    it("should not throw when both listResources and getResource are provided", () => {
        const listResources = async () => [];
        const getResource = async () => ({
            contents: [
                {
                    uri: "https://resource",
                    mimeType: "text/plain",
                    text: "test",
                },
            ],
        });
        expect(() => {
            (0, resource_validators_1.validateResourceSource)(listResources, getResource);
        }).not.toThrow();
    });
    it("should not throw when neither listResources nor getResource are provided", () => {
        const listResources = undefined;
        const getResource = undefined;
        expect(() => {
            (0, resource_validators_1.validateResourceSource)(listResources, getResource);
        }).not.toThrow();
    });
});
describe("validateResource", () => {
    it("should throw when resource is missing uri field", () => {
        const resource = {
            name: "Test Resource",
        };
        expect(() => {
            (0, resource_validators_1.validateResource)(resource);
        }).toThrow("Resource must have a 'uri' field");
    });
    it("should throw when resource is missing name field", () => {
        const resource = {
            uri: "https://resource",
        };
        expect(() => {
            (0, resource_validators_1.validateResource)(resource);
        }).toThrow("Resource with URI 'https://resource' must have a 'name' field");
    });
    it("should not throw when resource has both uri and name fields", () => {
        const resource = {
            uri: "https://resource",
            name: "Test Resource",
        };
        expect(() => {
            (0, resource_validators_1.validateResource)(resource);
        }).not.toThrow();
    });
    it("should not throw when resource has uri, name, and optional fields", () => {
        const resource = {
            uri: "https://resource",
            name: "Test Resource",
            description: "A test resource",
            mimeType: "text/plain",
        };
        expect(() => {
            (0, resource_validators_1.validateResource)(resource);
        }).not.toThrow();
    });
});
