// @tambo-ai/react/dist/util/tool-caller.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleToolCall = void 0;
const registry_1 = require("./registry");
/**
 * Process a message from the thread, invoking the appropriate tool and returning the result.
 * @param toolCallRequest - The message to handle
 * @param toolRegistry - The tool registry
 * @returns The result of the tool call along with the tool definition
 */
const handleToolCall = async (toolCallRequest, toolRegistry, onCallUnregisteredTool) => {
    if (!toolCallRequest?.toolName) {
        throw new Error("Tool name is required");
    }
    try {
        const { tool, tamboTool } = findTool(toolCallRequest.toolName, toolRegistry);
        if (!tool) {
            if (onCallUnregisteredTool) {
                const result = await onCallUnregisteredTool(toolCallRequest.toolName, toolCallRequest.parameters);
                return {
                    result,
                };
            }
            throw new Error(`Tool ${toolCallRequest.toolName} not found in registry`);
        }
        return {
            result: await runToolChoice(toolCallRequest, tool),
            tamboTool,
        };
    }
    catch (error) {
        console.error("Error in calling tool: ", error);
        return {
            result: `When attempting to call tool ${toolCallRequest.toolName} the following error occurred: ${error}. Explain to the user that the tool call failed and try again if needed.`,
            error: error instanceof Error ? error.message : "Unknown error",
        };
    }
};
exports.handleToolCall = handleToolCall;
const findTool = (toolName, toolRegistry) => {
    const registryTool = toolRegistry[toolName];
    if (!registryTool) {
        return { tool: null, tamboTool: null };
    }
    const contextTool = (0, registry_1.mapTamboToolToContextTool)(registryTool);
    return {
        tool: {
            getComponentContext: registryTool.tool,
            definition: contextTool,
        },
        tamboTool: registryTool,
    };
};
const runToolChoice = async (toolCallRequest, tool) => {
    const parameters = toolCallRequest.parameters ?? [];
    // Reconstruct the object from parameter name-value pairs
    const inputObject = Object.fromEntries(parameters.map((param) => [param.parameterName, param.parameterValue]));
    return await tool.getComponentContext(inputObject);
};
