// @tambo-ai/react/dist/v1/hooks/use-tambo-v1-suggestions.js
"use strict";
"use client";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTamboV1Suggestions = useTamboV1Suggestions;
const react_1 = require("react");
const tambo_client_provider_1 = require("../../providers/tambo-client-provider");
const tambo_registry_provider_1 = require("../../providers/tambo-registry-provider");
const registry_1 = require("../../util/registry");
const react_query_hooks_1 = require("../../hooks/react-query-hooks");
const query_utils_1 = require("../../util/query-utils");
const use_tambo_v1_1 = require("./use-tambo-v1");
const use_tambo_v1_thread_input_1 = require("./use-tambo-v1-thread-input");
function normalizeSuggestionGenerateResponse(response) {
    if (response === undefined) {
        return [];
    }
    if (Array.isArray(response)) {
        return response;
    }
    throw new Error("Unexpected suggestions.generate response shape; expected an array");
}
/**
 * Hook for managing Tambo AI suggestions in a v1 thread.
 *
 * Automatically generates suggestions when the latest message is from the assistant
 * and the thread is idle (not streaming).
 *
 * Uses the shared thread input context, so accepting a suggestion without submitting
 * will automatically update the input field value for any component using
 * useTamboV1ThreadInput.
 * @param options - Configuration options for suggestion generation
 * @returns Object containing suggestions state and control functions
 * @example
 * ```tsx
 * function SuggestionsPanel() {
 *   const { suggestions, accept, isPending } = useTamboV1Suggestions();
 *
 *   return (
 *     <div>
 *       {suggestions.map(suggestion => (
 *         <button
 *           key={suggestion.id}
 *           onClick={() => accept({ suggestion })}
 *           disabled={isPending}
 *         >
 *           {suggestion.suggestion}
 *         </button>
 *       ))}
 *     </div>
 *   );
 * }
 *
 * function ChatInput() {
 *   // This input will automatically show the accepted suggestion
 *   const { value, setValue, submit } = useTamboV1ThreadInput();
 *   return <input value={value} onChange={e => setValue(e.target.value)} />;
 * }
 * ```
 */
function useTamboV1Suggestions(options = {}) {
    const { maxSuggestions = 3 } = options;
    // Use shared thread input context
    const { setValue, submit, threadId } = (0, use_tambo_v1_thread_input_1.useTamboV1ThreadInput)();
    const { messages, isIdle } = (0, use_tambo_v1_1.useTamboV1)(threadId);
    const tamboClient = (0, tambo_client_provider_1.useTamboClient)();
    const { componentList, toolRegistry, componentToolAssociations } = (0, tambo_registry_provider_1.useTamboRegistry)();
    const [selectedSuggestionId, setSelectedSuggestionId] = (0, react_1.useState)(null);
    const latestMessage = messages[messages.length - 1];
    const isLatestFromAssistant = latestMessage?.role === "assistant";
    const latestMessageId = latestMessage?.id;
    // Reset selected suggestion when the message changes
    (0, react_1.useEffect)(() => {
        setSelectedSuggestionId(null);
    }, [latestMessageId]);
    const shouldGenerateSuggestions = latestMessageId && isLatestFromAssistant && isIdle && threadId;
    // Use React Query to fetch suggestions when conditions are met
    const suggestionsResult = (0, react_query_hooks_1.useTamboQuery)({
        queryKey: [
            "v1-suggestions",
            shouldGenerateSuggestions ? latestMessageId : null,
        ],
        queryFn: async () => {
            if (!shouldGenerateSuggestions || !threadId || !latestMessageId) {
                return [];
            }
            // Get registered components from the registry
            const components = (0, registry_1.getAvailableComponents)(componentList, toolRegistry, componentToolAssociations);
            const response = await tamboClient.beta.threads.suggestions.generate(latestMessageId, {
                id: threadId,
                maxSuggestions,
                availableComponents: components,
            });
            return normalizeSuggestionGenerateResponse(response);
        },
        enabled: Boolean(shouldGenerateSuggestions),
        refetchOnWindowFocus: false,
        refetchOnReconnect: false,
        retry: false,
    });
    // Accept suggestion mutation
    const acceptMutationState = (0, react_query_hooks_1.useTamboMutation)({
        mutationFn: async ({ suggestion, shouldSubmit = false }) => {
            const text = suggestion.detailedSuggestion?.trim();
            if (!text) {
                throw new Error("Suggestion has no detailed content");
            }
            if (shouldSubmit) {
                // Set the value first, then submit
                setValue(text);
                await submit();
            }
            else {
                // Just update the shared input value
                setValue(text);
            }
            setSelectedSuggestionId(suggestion.id);
        },
    });
    // Generate suggestions mutation (for manual refresh)
    const generateMutationState = (0, react_query_hooks_1.useTamboMutation)({
        mutationFn: async (abortController) => {
            if (!shouldGenerateSuggestions || !threadId || !latestMessageId) {
                return [];
            }
            const components = (0, registry_1.getAvailableComponents)(componentList, toolRegistry, componentToolAssociations);
            const response = await tamboClient.beta.threads.suggestions.generate(latestMessageId, {
                id: threadId,
                maxSuggestions,
                availableComponents: components,
            }, { signal: abortController.signal });
            return normalizeSuggestionGenerateResponse(response);
        },
        retry: false,
    });
    // Use query data if available, otherwise use mutation data
    const suggestions = isLatestFromAssistant
        ? (suggestionsResult.data ?? generateMutationState.data ?? [])
        : [];
    const accept = (0, react_1.useCallback)(async (acceptOptions) => {
        await acceptMutationState.mutateAsync(acceptOptions);
    }, [acceptMutationState]);
    return {
        suggestions,
        accept,
        selectedSuggestionId,
        acceptResult: acceptMutationState,
        generateResult: generateMutationState,
        suggestionsResult,
        ...(0, query_utils_1.combineMutationResults)(acceptMutationState, generateMutationState),
    };
}
