// @tambo-ai/react/dist/v1/hooks/use-tambo-v1.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = require("@testing-library/react");
const react_2 = __importDefault(require("react"));
const tambo_client_provider_1 = require("../../providers/tambo-client-provider");
const tambo_registry_provider_1 = require("../../providers/tambo-registry-provider");
const tambo_v1_stream_context_1 = require("../providers/tambo-v1-stream-context");
const use_tambo_v1_1 = require("./use-tambo-v1");
jest.mock("../../providers/tambo-client-provider", () => ({
    useTamboClient: jest.fn(),
}));
describe("useTamboV1", () => {
    const mockTamboClient = {
        apiKey: "",
        threads: {},
    };
    const mockRegistry = {
        componentList: {},
        toolRegistry: {},
        componentToolAssociations: {},
        mcpServerInfos: [],
        resources: [],
        resourceSource: null,
        registerComponent: jest.fn(),
        registerTool: jest.fn(),
        registerTools: jest.fn(),
        addToolAssociation: jest.fn(),
        registerMcpServer: jest.fn(),
        registerMcpServers: jest.fn(),
        registerResource: jest.fn(),
        registerResources: jest.fn(),
        registerResourceSource: jest.fn(),
    };
    function TestWrapper({ children }) {
        return (react_2.default.createElement(tambo_registry_provider_1.TamboRegistryContext.Provider, { value: mockRegistry },
            react_2.default.createElement(tambo_v1_stream_context_1.TamboV1StreamProvider, null, children)));
    }
    beforeEach(() => {
        jest.mocked(tambo_client_provider_1.useTamboClient).mockReturnValue(mockTamboClient);
        jest.clearAllMocks();
    });
    it("returns client from useTamboClient", () => {
        const { result } = (0, react_1.renderHook)(() => (0, use_tambo_v1_1.useTamboV1)(), {
            wrapper: TestWrapper,
        });
        expect(result.current.client).toBe(mockTamboClient);
    });
    it("returns registry functions", () => {
        const { result } = (0, react_1.renderHook)(() => (0, use_tambo_v1_1.useTamboV1)(), {
            wrapper: TestWrapper,
        });
        expect(result.current.registerComponent).toBe(mockRegistry.registerComponent);
        expect(result.current.registerTool).toBe(mockRegistry.registerTool);
        expect(result.current.registerTools).toBe(mockRegistry.registerTools);
        expect(result.current.componentList).toBe(mockRegistry.componentList);
        expect(result.current.toolRegistry).toBe(mockRegistry.toolRegistry);
    });
    it("returns undefined thread when no threadId provided", () => {
        const { result } = (0, react_1.renderHook)(() => (0, use_tambo_v1_1.useTamboV1)(), {
            wrapper: TestWrapper,
        });
        expect(result.current.thread).toBeUndefined();
        expect(result.current.messages).toEqual([]);
    });
    it("returns thread state when threadId provided", () => {
        const { result } = (0, react_1.renderHook)(() => (0, use_tambo_v1_1.useTamboV1)("thread_123"), {
            wrapper: TestWrapper,
        });
        // Initialize thread first
        (0, react_1.act)(() => {
            result.current.initThread("thread_123");
        });
        expect(result.current.thread).toBeDefined();
        expect(result.current.thread?.thread.id).toBe("thread_123");
        expect(result.current.messages).toEqual([]);
    });
    it("returns default streaming state when thread not loaded", () => {
        const { result } = (0, react_1.renderHook)(() => (0, use_tambo_v1_1.useTamboV1)(), {
            wrapper: TestWrapper,
        });
        expect(result.current.streamingState).toEqual({ status: "idle" });
        expect(result.current.isIdle).toBe(true);
        expect(result.current.isStreaming).toBe(false);
        expect(result.current.isWaiting).toBe(false);
    });
    it("returns thread streaming state when thread loaded", () => {
        const { result } = (0, react_1.renderHook)(() => (0, use_tambo_v1_1.useTamboV1)("thread_123"), {
            wrapper: TestWrapper,
        });
        // Initialize thread first
        (0, react_1.act)(() => {
            result.current.initThread("thread_123");
        });
        expect(result.current.streamingState.status).toBe("idle");
        expect(result.current.isIdle).toBe(true);
        expect(result.current.isStreaming).toBe(false);
        expect(result.current.isWaiting).toBe(false);
    });
    it("provides dispatch function for advanced usage", () => {
        const { result } = (0, react_1.renderHook)(() => (0, use_tambo_v1_1.useTamboV1)(), {
            wrapper: TestWrapper,
        });
        expect(typeof result.current.dispatch).toBe("function");
    });
    it("provides thread management functions", () => {
        const { result } = (0, react_1.renderHook)(() => (0, use_tambo_v1_1.useTamboV1)(), {
            wrapper: TestWrapper,
        });
        expect(typeof result.current.initThread).toBe("function");
        expect(typeof result.current.switchThread).toBe("function");
        expect(typeof result.current.startNewThread).toBe("function");
    });
    it("initializes and switches threads", () => {
        const { result } = (0, react_1.renderHook)(() => (0, use_tambo_v1_1.useTamboV1)(), {
            wrapper: TestWrapper,
        });
        // Initially no current thread
        expect(result.current.currentThreadId).toBeNull();
        expect(result.current.thread).toBeUndefined();
        // Initialize a new thread
        (0, react_1.act)(() => {
            result.current.initThread("new_thread_1");
        });
        // Switch to the new thread
        (0, react_1.act)(() => {
            result.current.switchThread("new_thread_1");
        });
        expect(result.current.currentThreadId).toBe("new_thread_1");
    });
    it("starts new thread with generated ID", () => {
        const { result } = (0, react_1.renderHook)(() => (0, use_tambo_v1_1.useTamboV1)(), {
            wrapper: TestWrapper,
        });
        let newThreadId;
        (0, react_1.act)(() => {
            newThreadId = result.current.startNewThread();
        });
        expect(newThreadId).toMatch(/^temp_/);
        expect(result.current.currentThreadId).toBe(newThreadId);
        expect(result.current.thread).toBeDefined();
    });
    it("uses current thread when no threadId argument provided", () => {
        const { result } = (0, react_1.renderHook)(() => (0, use_tambo_v1_1.useTamboV1)(), {
            wrapper: TestWrapper,
        });
        // Initialize and switch to a thread
        (0, react_1.act)(() => {
            result.current.initThread("thread_123");
            result.current.switchThread("thread_123");
        });
        // Should use current thread from context (thread_123)
        expect(result.current.currentThreadId).toBe("thread_123");
        expect(result.current.thread?.thread.id).toBe("thread_123");
    });
});
