// @tambo-ai/react/dist/v1/hooks/use-tambo-v1.js
"use strict";
"use client";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTamboV1 = useTamboV1;
/**
 * useTamboV1 - Main Hook for v1 API
 *
 * Combines all v1 contexts into a single hook for convenient access
 * to thread state, streaming status, registry, and client.
 */
const react_1 = require("react");
const tambo_client_provider_1 = require("../../providers/tambo-client-provider");
const tambo_registry_provider_1 = require("../../providers/tambo-registry-provider");
const tambo_v1_stream_context_1 = require("../providers/tambo-v1-stream-context");
/**
 * Main hook for accessing Tambo v1 functionality.
 *
 * Combines thread state, streaming status, registry, and client
 * into a single convenient hook.
 * @param threadId - Optional thread ID to get state for
 * @returns Combined v1 context with thread state and utilities
 * @example
 * ```tsx
 * function ChatInterface() {
 *   const {
 *     thread,
 *     messages,
 *     isStreaming,
 *     registerComponent,
 *   } = useTamboV1('thread_123');
 *
 *   return (
 *     <div>
 *       {messages.map(msg => <Message key={msg.id} message={msg} />)}
 *       {isStreaming && <LoadingIndicator />}
 *     </div>
 *   );
 * }
 * ```
 */
function useTamboV1(threadId) {
    const client = (0, tambo_client_provider_1.useTamboClient)();
    const streamState = (0, tambo_v1_stream_context_1.useStreamState)();
    const dispatch = (0, tambo_v1_stream_context_1.useStreamDispatch)();
    const registry = (0, react_1.useContext)(tambo_registry_provider_1.TamboRegistryContext);
    const threadManagement = (0, tambo_v1_stream_context_1.useThreadManagement)();
    // Get thread state for the given threadId (or current thread if not specified)
    const effectiveThreadId = threadId ?? streamState.currentThreadId;
    const threadState = effectiveThreadId
        ? streamState.threadMap[effectiveThreadId]
        : undefined;
    // Memoize the return object to prevent unnecessary re-renders
    return (0, react_1.useMemo)(() => {
        const thread = threadState;
        const messages = thread?.thread.messages ?? [];
        const streamingState = thread?.streaming ?? {
            status: "idle",
        };
        return {
            client,
            thread,
            messages,
            streamingState,
            isStreaming: streamingState.status === "streaming",
            isWaiting: streamingState.status === "waiting",
            isIdle: streamingState.status === "idle",
            registerComponent: registry.registerComponent,
            registerTool: registry.registerTool,
            registerTools: registry.registerTools,
            componentList: registry.componentList,
            toolRegistry: registry.toolRegistry,
            currentThreadId: streamState.currentThreadId,
            initThread: threadManagement.initThread,
            switchThread: threadManagement.switchThread,
            startNewThread: threadManagement.startNewThread,
            dispatch,
        };
    }, [
        client,
        threadState,
        registry,
        streamState.currentThreadId,
        threadManagement,
        dispatch,
    ]);
}
