// @tambo-ai/typescript-sdk/src/resources/beta/projects/api-key.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import { APIResource } from '../../../core/resource';
import { APIPromise } from '../../../core/api-promise';
import { RequestOptions } from '../../../internal/request-options';
import { path } from '../../../internal/utils/path';

export class APIKey extends APIResource {
  /**
   * @example
   * ```ts
   * const apiKeys = await client.beta.projects.apiKey.list(
   *   'id',
   * );
   * ```
   */
  list(id: string, options?: RequestOptions): APIPromise<APIKeyListResponse> {
    return this._client.get(path`/projects/${id}/api-keys`, options);
  }

  /**
   * @example
   * ```ts
   * const apiKey = await client.beta.projects.apiKey.delete(
   *   'apiKeyId',
   *   { id: 'id' },
   * );
   * ```
   */
  delete(
    apiKeyID: string,
    params: APIKeyDeleteParams,
    options?: RequestOptions,
  ): APIPromise<APIKeyDeleteResponse> {
    const { id } = params;
    return this._client.delete(path`/projects/${id}/api-key/${apiKeyID}`, options);
  }
}

export type APIKeyListResponse = Array<APIKeyListResponse.APIKeyListResponseItem>;

export namespace APIKeyListResponse {
  export interface APIKeyListResponseItem {
    id?: string;

    created?: string;

    createdByUserId?: string;

    lastUsed?: string;

    name?: string;

    partiallyHiddenKey?: string;
  }
}

export type APIKeyDeleteResponse = boolean;

export interface APIKeyDeleteParams {
  id: string;
}

export declare namespace APIKey {
  export {
    type APIKeyListResponse as APIKeyListResponse,
    type APIKeyDeleteResponse as APIKeyDeleteResponse,
    type APIKeyDeleteParams as APIKeyDeleteParams,
  };
}
