// @tambo-ai/typescript-sdk/src/resources/beta/projects/projects.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import { APIResource } from '../../../core/resource';
import * as APIKeyAPI from './api-key';
import { APIKey, APIKeyDeleteParams, APIKeyDeleteResponse, APIKeyListResponse } from './api-key';
import * as ProviderKeyAPI from './provider-key';
import {
  ProviderKey,
  ProviderKeyDeleteParams,
  ProviderKeyDeleteResponse,
  ProviderKeyListResponse,
} from './provider-key';
import { APIPromise } from '../../../core/api-promise';
import { RequestOptions } from '../../../internal/request-options';
import { path } from '../../../internal/utils/path';

export class Projects extends APIResource {
  apiKey: APIKeyAPI.APIKey = new APIKeyAPI.APIKey(this._client);
  providerKey: ProviderKeyAPI.ProviderKey = new ProviderKeyAPI.ProviderKey(this._client);

  /**
   * @example
   * ```ts
   * const project = await client.beta.projects.retrieve('id');
   * ```
   */
  retrieve(id: string, options?: RequestOptions): APIPromise<ProjectRetrieveResponse> {
    return this._client.get(path`/projects/${id}`, options);
  }

  /**
   * @example
   * ```ts
   * const project = await client.beta.projects.delete('id');
   * ```
   */
  delete(id: string, options?: RequestOptions): APIPromise<ProjectDeleteResponse> {
    return this._client.delete(path`/projects/${id}`, options);
  }

  /**
   * @example
   * ```ts
   * const response = await client.beta.projects.getCurrent();
   * ```
   */
  getCurrent(options?: RequestOptions): APIPromise<ProjectGetCurrentResponse> {
    return this._client.get('/projects', options);
  }
}

export interface ProjectRetrieveResponse {
  id: string;

  isTokenRequired: boolean;

  name: string;

  providerType: 'llm' | 'agent';

  userId: string;

  agentName?: string;

  agentProviderType?: 'ag-ui' | 'mastra' | 'crewai' | 'llamaindex' | 'langgraph' | 'pydanticai';

  agentUrl?: string;

  allowSystemPromptOverride?: boolean;

  customInstructions?: string;

  customLlmBaseURL?: string;

  customLlmModelName?: string;

  customLlmParameters?: unknown;

  defaultLlmModelName?: string;

  defaultLlmProviderName?: string;

  maxInputTokens?: number | null;
}

export type ProjectDeleteResponse = boolean;

export interface ProjectGetCurrentResponse {
  id: string;

  isTokenRequired: boolean;

  name: string;

  providerType: 'llm' | 'agent';

  userId: string;

  agentName?: string;

  agentProviderType?: 'ag-ui' | 'mastra' | 'crewai' | 'llamaindex' | 'langgraph' | 'pydanticai';

  agentUrl?: string;

  allowSystemPromptOverride?: boolean;

  customInstructions?: string;

  customLlmBaseURL?: string;

  customLlmModelName?: string;

  customLlmParameters?: unknown;

  defaultLlmModelName?: string;

  defaultLlmProviderName?: string;

  maxInputTokens?: number | null;
}

Projects.APIKey = APIKey;
Projects.ProviderKey = ProviderKey;

export declare namespace Projects {
  export {
    type ProjectRetrieveResponse as ProjectRetrieveResponse,
    type ProjectDeleteResponse as ProjectDeleteResponse,
    type ProjectGetCurrentResponse as ProjectGetCurrentResponse,
  };

  export {
    APIKey as APIKey,
    type APIKeyListResponse as APIKeyListResponse,
    type APIKeyDeleteResponse as APIKeyDeleteResponse,
    type APIKeyDeleteParams as APIKeyDeleteParams,
  };

  export {
    ProviderKey as ProviderKey,
    type ProviderKeyListResponse as ProviderKeyListResponse,
    type ProviderKeyDeleteResponse as ProviderKeyDeleteResponse,
    type ProviderKeyDeleteParams as ProviderKeyDeleteParams,
  };
}
