// @tambo-ai/typescript-sdk/src/resources/beta/projects/provider-key.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import { APIResource } from '../../../core/resource';
import { APIPromise } from '../../../core/api-promise';
import { RequestOptions } from '../../../internal/request-options';
import { path } from '../../../internal/utils/path';

export class ProviderKey extends APIResource {
  /**
   * @example
   * ```ts
   * const providerKeys =
   *   await client.beta.projects.providerKey.list('id');
   * ```
   */
  list(id: string, options?: RequestOptions): APIPromise<ProviderKeyListResponse> {
    return this._client.get(path`/projects/${id}/provider-keys`, options);
  }

  /**
   * @example
   * ```ts
   * const providerKey =
   *   await client.beta.projects.providerKey.delete(
   *     'providerKeyId',
   *     { id: 'id' },
   *   );
   * ```
   */
  delete(
    providerKeyID: string,
    params: ProviderKeyDeleteParams,
    options?: RequestOptions,
  ): APIPromise<ProviderKeyDeleteResponse> {
    const { id } = params;
    return this._client.delete(path`/projects/${id}/provider-key/${providerKeyID}`, options);
  }
}

export type ProviderKeyListResponse = Array<ProviderKeyListResponse.ProviderKeyListResponseItem>;

export namespace ProviderKeyListResponse {
  export interface ProviderKeyListResponseItem {
    id?: string;

    partiallyHiddenKey?: string;

    providerKeyEncrypted?: string;

    providerName?: string;
  }
}

export interface ProviderKeyDeleteResponse {
  id: string;

  name: string;
}

export interface ProviderKeyDeleteParams {
  id: string;
}

export declare namespace ProviderKey {
  export {
    type ProviderKeyListResponse as ProviderKeyListResponse,
    type ProviderKeyDeleteResponse as ProviderKeyDeleteResponse,
    type ProviderKeyDeleteParams as ProviderKeyDeleteParams,
  };
}
