// @tambo-ai/typescript-sdk/src/lib/advance-stream.ts
import { type TamboAI } from '../client';
import { APIPromise } from '../core/api-promise';
import { RequestOptions } from '../internal/request-options';
import { ThreadAdvanceByIDParams, ThreadAdvanceResponse } from '../resources/beta';

export async function advanceStream(
  client: TamboAI,
  body: ThreadAdvanceByIDParams,
  threadId?: string,
  options?: RequestOptions,
): Promise<AsyncIterable<ThreadAdvanceResponse>> {
  const MAX_REQUEST_RETRIES = 2;
  let requestRetryCount = 0;

  while (true) {
    try {
      const responsePromise: APIPromise<ThreadAdvanceResponse> = client.post(
        `/threads${threadId ? `/${threadId}` : ''}/advancestream`,
        {
          body,
          ...options,
          headers: {
            ...options?.headers,
            Accept: 'text/event-stream',
          },
        },
      );
      const response = await responsePromise.asResponse();
      return handleStreamResponse<ThreadAdvanceResponse>(response);
    } catch (error) {
      if (requestRetryCount < MAX_REQUEST_RETRIES) {
        requestRetryCount++;
        console.warn(`Request failed, attempting retry ${requestRetryCount}/${MAX_REQUEST_RETRIES}`);
        continue;
      }
      throw error;
    }
  }
}

async function* handleStreamResponse<T>(response: Response): AsyncIterable<T> {
  const decoder = new TextDecoder();
  const MAX_CHUNK_RETRIES = 5;
  let chunkRetryCount = 0;
  const reader = (response.body as unknown as ReadableStream).getReader();

  while (true) {
    try {
      const { done, value: chunk } = await reader.read();
      if (done) break;

      const text = decoder.decode(chunk as Uint8Array);
      const messages = text.split('\n').filter((msg) => msg.trim());

      for (const msg of messages) {
        if (msg === 'data: DONE') {
          continue;
        }

        if (msg.startsWith('error: ')) {
          throw new Error(msg.slice(7));
        }

        const jsonStr = msg.startsWith('data: ') ? msg.slice(6) : msg;

        if (!jsonStr) {
          continue;
        }

        try {
          yield JSON.parse(jsonStr) as T;
          chunkRetryCount = 0;
        } catch (e) {
          if (chunkRetryCount < MAX_CHUNK_RETRIES) {
            chunkRetryCount++;
            console.warn(
              `Failed to parse JSON chunk, skipping. Attempt ${chunkRetryCount}/${MAX_CHUNK_RETRIES}`,
            );
            continue;
          }
          throw new Error('Failed to parse JSON after multiple chunks.');
        }
      }
    } catch (error) {
      reader.releaseLock();
      throw error;
    }
  }
}
