// @tambo-ai/react/dist/hoc/with-tambo-interactable.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = require("@testing-library/react");
const react_2 = __importDefault(require("react"));
const v3_1 = require("zod/v3");
const use_current_message_1 = require("../hooks/use-current-message");
const tambo_interactable_provider_1 = require("../providers/tambo-interactable-provider");
const with_tambo_interactable_1 = require("./with-tambo-interactable");
// Create a consistent mock implementation
const mockAddInteractableComponent = jest.fn(() => "mock-id-123");
const mockUpdateInteractableComponentProps = jest.fn();
const mockGetInteractableComponent = jest.fn(() => null);
const mockRemoveInteractableComponent = jest.fn();
const mockGetInteractableComponentsByName = jest.fn(() => []);
const mockClearAllInteractableComponents = jest.fn();
// Mock the interactable provider
jest.mock("../providers/tambo-interactable-provider", () => ({
    TamboInteractableProvider: ({ children }) => (react_2.default.createElement("div", { "data-testid": "interactable-provider" }, children)),
    useTamboInteractable: () => ({
        addInteractableComponent: mockAddInteractableComponent,
        updateInteractableComponentProps: mockUpdateInteractableComponentProps,
        getInteractableComponent: mockGetInteractableComponent,
        removeInteractableComponent: mockRemoveInteractableComponent,
        getInteractableComponentsByName: mockGetInteractableComponentsByName,
        clearAllInteractableComponents: mockClearAllInteractableComponents,
        interactableComponents: [],
    }),
}));
describe("withTamboInteractable", () => {
    const TestComponent = ({ title, count }) => (react_2.default.createElement("div", null,
        react_2.default.createElement("h1", { "data-testid": "title" }, title),
        count !== undefined && react_2.default.createElement("span", { "data-testid": "count" }, count)));
    const testSchema = v3_1.z.object({
        title: v3_1.z.string(),
        count: v3_1.z.number().optional(),
    });
    beforeEach(() => {
        jest.clearAllMocks();
        // Reset mock implementations
        mockAddInteractableComponent.mockReturnValue("mock-id-123");
        mockGetInteractableComponent.mockReturnValue(null);
        mockGetInteractableComponentsByName.mockReturnValue([]);
    });
    it("should render the wrapped component with provided props", () => {
        const InteractableComponent = (0, with_tambo_interactable_1.withTamboInteractable)(TestComponent, {
            componentName: "TestComponent",
            description: "A test component",
            propsSchema: testSchema,
        });
        (0, react_1.render)(react_2.default.createElement(tambo_interactable_provider_1.TamboInteractableProvider, null,
            react_2.default.createElement(InteractableComponent, { title: "Hello", count: 42 })));
        expect(react_1.screen.getByTestId("title")).toHaveTextContent("Hello");
        expect(react_1.screen.getByTestId("count")).toHaveTextContent("42");
    });
    it("should set displayName correctly", () => {
        const InteractableComponent = (0, with_tambo_interactable_1.withTamboInteractable)(TestComponent, {
            componentName: "TestComponent",
            description: "A test component",
        });
        expect(InteractableComponent.displayName).toBe("withTamboInteractable(TestComponent)");
    });
    it("should handle components without displayName", () => {
        const AnonymousComponent = () => react_2.default.createElement("div", null, "Anonymous");
        const InteractableComponent = (0, with_tambo_interactable_1.withTamboInteractable)(AnonymousComponent, {
            componentName: "AnonymousComponent",
            description: "An anonymous component",
        });
        // Anonymous components get their name from the function name
        expect(InteractableComponent.displayName).toBe("withTamboInteractable(AnonymousComponent)");
    });
    it("should provide TamboMessageProvider with interactable metadata", async () => {
        // Child component that uses the context
        const ContextConsumer = () => {
            const component = (0, use_current_message_1.useTamboCurrentComponent)();
            return (react_2.default.createElement("div", null,
                react_2.default.createElement("span", { "data-testid": "interactable-id" }, component?.interactableId),
                react_2.default.createElement("span", { "data-testid": "component-name" }, component?.componentName),
                react_2.default.createElement("span", { "data-testid": "description" }, component?.description)));
        };
        // Wrap TestComponent to include ContextConsumer
        const TestComponentWithConsumer = (props) => (react_2.default.createElement("div", null,
            react_2.default.createElement(TestComponent, { ...props }),
            react_2.default.createElement(ContextConsumer, null)));
        const InteractableWithConsumer = (0, with_tambo_interactable_1.withTamboInteractable)(TestComponentWithConsumer, {
            componentName: "TestComponent",
            description: "A test component",
            propsSchema: testSchema,
        });
        (0, react_1.render)(react_2.default.createElement(tambo_interactable_provider_1.TamboInteractableProvider, null,
            react_2.default.createElement(InteractableWithConsumer, { title: "Hello" })));
        // Wait for the component to register and render
        await (0, react_1.waitFor)(() => {
            expect(react_1.screen.getByTestId("interactable-id")).toHaveTextContent("mock-id-123");
        });
        expect(react_1.screen.getByTestId("component-name")).toHaveTextContent("TestComponent");
        expect(react_1.screen.getByTestId("description")).toHaveTextContent("A test component");
    });
    it("should create minimal message with correct structure", async () => {
        // Child component that checks the message structure
        const MessageChecker = () => {
            const component = (0, use_current_message_1.useTamboCurrentComponent)();
            return (react_2.default.createElement("div", null,
                react_2.default.createElement("span", { "data-testid": "has-component-name" }, component?.componentName ? "yes" : "no"),
                react_2.default.createElement("span", { "data-testid": "has-props" }, component?.props ? "yes" : "no"),
                react_2.default.createElement("span", { "data-testid": "has-interactable-id" }, component?.interactableId ? "yes" : "no"),
                react_2.default.createElement("span", { "data-testid": "has-description" }, component?.description ? "yes" : "no")));
        };
        const TestComponentWithChecker = (props) => (react_2.default.createElement("div", null,
            react_2.default.createElement(TestComponent, { ...props }),
            react_2.default.createElement(MessageChecker, null)));
        const InteractableWithChecker = (0, with_tambo_interactable_1.withTamboInteractable)(TestComponentWithChecker, {
            componentName: "TestComponent",
            description: "A test component",
            propsSchema: testSchema,
        });
        (0, react_1.render)(react_2.default.createElement(tambo_interactable_provider_1.TamboInteractableProvider, null,
            react_2.default.createElement(InteractableWithChecker, { title: "Hello", count: 42 })));
        await (0, react_1.waitFor)(() => {
            expect(react_1.screen.getByTestId("has-component-name")).toHaveTextContent("yes");
        });
        expect(react_1.screen.getByTestId("has-props")).toHaveTextContent("yes");
        expect(react_1.screen.getByTestId("has-interactable-id")).toHaveTextContent("yes");
        expect(react_1.screen.getByTestId("has-description")).toHaveTextContent("yes");
    });
    it("should pass through all component props correctly", () => {
        const InteractableComponent = (0, with_tambo_interactable_1.withTamboInteractable)(TestComponent, {
            componentName: "TestComponent",
            description: "A test component",
            propsSchema: testSchema,
        });
        (0, react_1.render)(react_2.default.createElement(tambo_interactable_provider_1.TamboInteractableProvider, null,
            react_2.default.createElement(InteractableComponent, { title: "Test Title", count: 100 })));
        expect(react_1.screen.getByTestId("title")).toHaveTextContent("Test Title");
        expect(react_1.screen.getByTestId("count")).toHaveTextContent("100");
    });
    it("should filter out interactable-specific props from component props", () => {
        const ExtendedComponent = ({ title, onInteractableReady, onPropsUpdate, }) => (react_2.default.createElement("div", null,
            react_2.default.createElement("span", { "data-testid": "title" }, title),
            react_2.default.createElement("span", { "data-testid": "has-ready" }, onInteractableReady ? "yes" : "no"),
            react_2.default.createElement("span", { "data-testid": "has-update" }, onPropsUpdate ? "yes" : "no")));
        const InteractableComponent = (0, with_tambo_interactable_1.withTamboInteractable)(ExtendedComponent, {
            componentName: "ExtendedComponent",
            description: "An extended component",
        });
        const mockReady = jest.fn();
        const mockUpdate = jest.fn();
        (0, react_1.render)(react_2.default.createElement(tambo_interactable_provider_1.TamboInteractableProvider, null,
            react_2.default.createElement(InteractableComponent, { title: "Test", onInteractableReady: mockReady, onPropsUpdate: mockUpdate })));
        expect(react_1.screen.getByTestId("title")).toHaveTextContent("Test");
        // These props should be filtered out
        expect(react_1.screen.getByTestId("has-ready")).toHaveTextContent("no");
        expect(react_1.screen.getByTestId("has-update")).toHaveTextContent("no");
    });
    it("should work without propsSchema", () => {
        const InteractableComponent = (0, with_tambo_interactable_1.withTamboInteractable)(TestComponent, {
            componentName: "TestComponent",
            description: "A test component",
        });
        (0, react_1.render)(react_2.default.createElement(tambo_interactable_provider_1.TamboInteractableProvider, null,
            react_2.default.createElement(InteractableComponent, { title: "No Schema" })));
        expect(react_1.screen.getByTestId("title")).toHaveTextContent("No Schema");
    });
    it("should render before interactable ID is set", () => {
        const InteractableComponent = (0, with_tambo_interactable_1.withTamboInteractable)(TestComponent, {
            componentName: "TestComponent",
            description: "A test component",
        });
        const { container } = (0, react_1.render)(react_2.default.createElement(tambo_interactable_provider_1.TamboInteractableProvider, null,
            react_2.default.createElement(InteractableComponent, { title: "Early Render" })));
        // Component should render even before ID is available
        expect(container.querySelector('[data-testid="title"]')).toBeInTheDocument();
    });
    it("should handle null/undefined values in props", () => {
        const NullableComponent = ({ title, optional, nullable, }) => (react_2.default.createElement("div", null,
            react_2.default.createElement("span", { "data-testid": "title" }, title),
            react_2.default.createElement("span", { "data-testid": "optional" }, optional ?? "undefined"),
            react_2.default.createElement("span", { "data-testid": "nullable" }, nullable ?? "null")));
        const InteractableComponent = (0, with_tambo_interactable_1.withTamboInteractable)(NullableComponent, {
            componentName: "NullableComponent",
            description: "A component with nullable props",
        });
        (0, react_1.render)(react_2.default.createElement(tambo_interactable_provider_1.TamboInteractableProvider, null,
            react_2.default.createElement(InteractableComponent, { title: "Test", nullable: null })));
        expect(react_1.screen.getByTestId("title")).toHaveTextContent("Test");
        expect(react_1.screen.getByTestId("optional")).toHaveTextContent("undefined");
        expect(react_1.screen.getByTestId("nullable")).toHaveTextContent("null");
    });
});
