// @tambo-ai/react/dist/testing/tools.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeRegistry = serializeRegistry;
exports.createMockTool = createMockTool;
const schema_1 = require("../schema");
const registry_1 = require("../util/registry");
/**
 * Serializes the registry for testing purposes.
 * Converts Standard Schema validators to JSON Schema format.
 * Uses the same logic as production code via mapTamboToolToContextTool.
 * @param mockRegistry - The registry to serialize
 * @returns The serialized registry with JSON Schema representations
 */
function serializeRegistry(mockRegistry) {
    return mockRegistry.map(({ component: _component, propsSchema, associatedTools, ...componentEntry }) => ({
        ...componentEntry,
        props: (0, schema_1.isStandardSchema)(propsSchema)
            ? (0, schema_1.safeSchemaToJsonSchema)(propsSchema)
            : propsSchema,
        contextTools: associatedTools?.map((tool) => (0, registry_1.mapTamboToolToContextTool)(tool)),
    }));
}
// Helper to create a minimal TamboTool for testing
/**
 * Creates a mock TamboTool with the given input schema for testing purposes.
 * Accepts either an inputSchema directly or an options object with inputSchema and outputSchema.
 * @param schemaOrOptions - The input schema or options object
 * @returns A mock TamboTool instance
 * @internal
 */
function createMockTool(schemaOrOptions) {
    // Check if options object was passed
    const hasInputSchemaProperty = (obj) => {
        return (typeof obj === "object" &&
            obj !== null &&
            "inputSchema" in obj &&
            obj.inputSchema != null);
    };
    if (hasInputSchemaProperty(schemaOrOptions)) {
        return {
            name: "mockTool",
            description: "A mock tool for testing",
            tool: jest.fn().mockImplementation(() => { }),
            inputSchema: schemaOrOptions.inputSchema,
            outputSchema: schemaOrOptions.outputSchema ?? {
                type: "object",
                properties: {},
            },
            ...(schemaOrOptions.maxCalls !== undefined
                ? { maxCalls: schemaOrOptions.maxCalls }
                : {}),
        };
    }
    return {
        name: "mockTool",
        description: "A mock tool for testing",
        tool: jest.fn().mockImplementation(() => { }),
        inputSchema: schemaOrOptions,
        outputSchema: { type: "object", properties: {} },
    };
}
